{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "45821872-0790-462c-b2cd-124e786bb4f0",
   "metadata": {},
   "source": [
    "# RNA STAR\n",
    "Here we perform sequence alignments on the input FASTQ files. This produces `.sam` files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e1cda9f-0f79-4d65-a85f-280d8d14b007",
   "metadata": {},
   "outputs": [],
   "source": [
    "!apt install rna-star"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d005624-710b-44b3-b79a-25a9069170f8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f32f7068-d6d3-496f-854e-3be01da270af",
   "metadata": {},
   "source": [
    "# sambamba\n",
    "Here we convert the `.sam` files to `.bam` files and sort them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "443e22f6-7b33-4d29-a86a-fc3fa445ea23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading package lists... Done\n",
      "Building dependency tree... Done\n",
      "Reading state information... Done\n",
      "The following additional packages will be installed:\n",
      "  libllvm11 libphobos2-ldc-shared98\n",
      "The following NEW packages will be installed:\n",
      "  libllvm11 libphobos2-ldc-shared98 sambamba\n",
      "0 upgraded, 3 newly installed, 0 to remove and 1 not upgraded.\n",
      "Need to get 21.8 MB of archives.\n",
      "After this operation, 92.9 MB of additional disk space will be used.\n",
      "Get:1 http://mirrors.ocf.berkeley.edu/ubuntu jammy/universe amd64 libllvm11 amd64 1:11.1.0-6 [19.6 MB]\n",
      "Get:2 http://mirrors.ocf.berkeley.edu/ubuntu jammy/universe amd64 libphobos2-ldc-shared98 amd64 1:1.28.0-1ubuntu1 [1,487 kB]\n",
      "Get:3 http://mirrors.ocf.berkeley.edu/ubuntu jammy/universe amd64 sambamba amd64 0.8.2+dfsg-2 [695 kB]\n",
      "Fetched 21.8 MB in 1s (21.4 MB/s)m\u001b[33m\n",
      "\n",
      "\u001b7\u001b[0;23r\u001b8\u001b[1ASelecting previously unselected package libllvm11:amd64.\n",
      "(Reading database ... 414490 files and directories currently installed.)Reading database ... 65%(Reading database ... 70%(Reading database ... 75%(Reading database ... 80%(Reading database ... 85%(Reading database ... 90%(Reading database ... 95%(Reading database ... 100%\n",
      "Preparing to unpack .../libllvm11_1%3a11.1.0-6_amd64.deb ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [  0%]\u001b[49m\u001b[39m [..........................................................] \u001b8\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [  8%]\u001b[49m\u001b[39m [####......................................................] \u001b8Unpacking libllvm11:amd64 (1:11.1.0-6) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 15%]\u001b[49m\u001b[39m [########..................................................] \u001b8Selecting previously unselected package libphobos2-ldc-shared98:amd64.\n",
      "Preparing to unpack .../libphobos2-ldc-shared98_1%3a1.28.0-1ubuntu1_amd64.deb ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 23%]\u001b[49m\u001b[39m [#############.............................................] \u001b8Unpacking libphobos2-ldc-shared98:amd64 (1:1.28.0-1ubuntu1) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 31%]\u001b[49m\u001b[39m [#################.........................................] \u001b8Selecting previously unselected package sambamba.\n",
      "Preparing to unpack .../sambamba_0.8.2+dfsg-2_amd64.deb ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 38%]\u001b[49m\u001b[39m [######################....................................] \u001b8Unpacking sambamba (0.8.2+dfsg-2) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 46%]\u001b[49m\u001b[39m [##########################................................] \u001b8Setting up libllvm11:amd64 (1:11.1.0-6) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 54%]\u001b[49m\u001b[39m [###############################...........................] \u001b8\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 62%]\u001b[49m\u001b[39m [###################################.......................] \u001b8Setting up libphobos2-ldc-shared98:amd64 (1:1.28.0-1ubuntu1) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 69%]\u001b[49m\u001b[39m [########################################..................] \u001b8\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 77%]\u001b[49m\u001b[39m [############################################..............] \u001b8Setting up sambamba (0.8.2+dfsg-2) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 85%]\u001b[49m\u001b[39m [#################################################.........] \u001b8\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 92%]\u001b[49m\u001b[39m [#####################################################.....] \u001b8Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
      "Processing triggers for man-db (2.10.2-1) ...\n",
      "\n",
      "\u001b7\u001b[0;24r\u001b8\u001b[1A\u001b[J"
     ]
    }
   ],
   "source": [
    "!# password.txt is a text file with your password.\n",
    "!# If this gives an error (probably because you don't have sudo permissions),\n",
    "!# try running apt install -y sambamba instead.\n",
    "!sudo -S apt install -y sambamba < password.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b017d401-4698-4d93-92b4-bb6f2dd47e3e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dir results/00MatrigelCoating/\n",
      "dir results/00PLLCoating/\n",
      "dir results/01agrin/\n",
      "dir results/02PanLaminin/\n",
      "dir results/03Laminin211/\n",
      "dir results/04CollagenI/\n",
      "dir results/_STARtmp/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# note that bash scripts don't make output in Jupyter notebooks until they finish (sad!)\n",
    "\n",
    "mkdir -p ./sorted_bam\n",
    "for dir in results/*/; do\n",
    "  echo dir $dir\n",
    "  if [ -d \"$dir\" ]; then\n",
    "    for subdir in \"$dir\"/*/; do\n",
    "      subdir_name=$(basename \"$subdir\")\n",
    "      if [ -f \"$subdir/Aligned.out.sam\" ]; then\n",
    "        sam_file=\"$subdir/Aligned.out.sam\"\n",
    "        group=\"${dir:9:3}\"\n",
    "        short=\"${subdir_name:0:12}\"\n",
    "        output=$group-$short\n",
    "        sambamba view -S --format=bam $sam_file | sambamba sort -n -o sorted_bam/${output}.bam /dev/stdin\n",
    "      fi\n",
    "    done\n",
    "  fi\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f683af31-b47d-4261-95d0-d20d54025711",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ran sambamba on results/01agrin/lib103916_L2-ds.d1b626e6358749c5aa8c7a900453bf55/results/01agrin/lib103916_L2-ds.d1b626e6358749c5aa8c7a900453bf55/Aligned.out.sam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ran sambamba on results/01agrin/lib103917_L1-ds.6cb7a56486a743aa97aef4b4492c71db/results/01agrin/lib103917_L1-ds.6cb7a56486a743aa97aef4b4492c71db/Aligned.out.sam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ran sambamba on results/01agrin/lib103916_L1-ds.cf006f61b73c426b81c8530ea73ed130/results/01agrin/lib103916_L1-ds.cf006f61b73c426b81c8530ea73ed130/Aligned.out.sam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_26428/558986621.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     17\u001b[0m             \u001b[0moutput\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf\"{group}-{short}\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m             \u001b[0;31m# run sambamba on the file and put it in sorted_bam\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m             \u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"sambamba view -S --format=bam {sam_file} | sambamba sort -n -o sorted_bam/{output}.bam /dev/stdin\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mshell\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcheck\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Ran sambamba on {subfolderpath}/{sam_file}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.10/subprocess.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    503\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0mPopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mpopenargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    504\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 505\u001b[0;31m             \u001b[0mstdout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommunicate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    506\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mTimeoutExpired\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    507\u001b[0m             \u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkill\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.10/subprocess.py\u001b[0m in \u001b[0;36mcommunicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1144\u001b[0m                 \u001b[0mstderr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1145\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1146\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1147\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1148\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.10/subprocess.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1207\u001b[0m             \u001b[0mendtime\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_time\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1208\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1209\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_wait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1210\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1211\u001b[0m             \u001b[0;31m# https://bugs.python.org/issue25942\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.10/subprocess.py\u001b[0m in \u001b[0;36m_wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1957\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1958\u001b[0m                             \u001b[0;32mbreak\u001b[0m  \u001b[0;31m# Another thread waited.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1959\u001b[0;31m                         \u001b[0;34m(\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msts\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_try_wait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1960\u001b[0m                         \u001b[0;31m# Check the pid and loop as waitpid has been known to\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1961\u001b[0m                         \u001b[0;31m# return 0 even without WNOHANG in odd situations.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.10/subprocess.py\u001b[0m in \u001b[0;36m_try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1915\u001b[0m             \u001b[0;34m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1916\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1917\u001b[0;31m                 \u001b[0;34m(\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msts\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwaitpid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwait_flags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1918\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mChildProcessError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1919\u001b[0m                 \u001b[0;31m# This happens if SIGCLD is set to be ignored or waiting\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import pathlib\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "pathlib.Path(\"sorted_bam\").mkdir(exist_ok=True)\n",
    "for folder in os.listdir(\"results\"):\n",
    "    folderpath = f\"results/{folder}\"\n",
    "    if not os.path.isdir(folderpath):\n",
    "        continue\n",
    "    \n",
    "    for subfolder in os.listdir(folderpath):\n",
    "        subfolderpath = f\"{folderpath}/{subfolder}\"\n",
    "        sam_file = f\"{subfolderpath}/Aligned.out.sam\"\n",
    "        if os.path.isfile(sam_file):\n",
    "            group = folder[9:12]\n",
    "            short = subfolder[0:13]\n",
    "            output = f\"{group}-{short}\"\n",
    "            # run sambamba on the file and put it in sorted_bam\n",
    "            subprocess.run(f\"sambamba view -S --format=bam {sam_file} | sambamba sort -n -o sorted_bam/{output}.bam /dev/stdin\", shell=True, check=True)\n",
    "            print(f\"Ran sambamba on {subfolderpath}/{sam_file}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d0d8c64-faac-4d28-8d94-3c469a12b556",
   "metadata": {},
   "source": [
    "# htseq-count\n",
    "Here we get sequence counts, and store them in CSV files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "35d9670c-54cb-4d2b-a26e-fa923bc6548d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sorting sorted_bam/1ag-lib103915_L1.bam ...\n",
      "sorting sorted_bam/1ag-lib103915_L2.bam ...\n",
      "sorting sorted_bam/1ag-lib103916_L1.bam ...\n",
      "sorting sorted_bam/1ag-lib103916_L2.bam ...\n",
      "sorting sorted_bam/1ag-lib103917_L1.bam ...\n",
      "sorting sorted_bam/1ag-lib103917_L2.bam ...\n",
      "output to htseqResults/1a_output.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2667906 alignment record pairs processed.\n",
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2700000 alignment record pairs processed.\n",
      "2800000 alignment record pairs processed.\n",
      "2900000 alignment record pairs processed.\n",
      "2933419 alignment record pairs processed.\n",
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2700000 alignment record pairs processed.\n",
      "2800000 alignment record pairs processed.\n",
      "2804493 alignment record pairs processed.\n",
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2700000 alignment record pairs processed.\n",
      "2800000 alignment record pairs processed.\n",
      "2876643 alignment record pairs processed.\n",
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2700000 alignment record pairs processed.\n",
      "2800000 alignment record pairs processed.\n",
      "2900000 alignment record pairs processed.\n",
      "2916191 alignment record pairs processed.\n",
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2700000 alignment record pairs processed.\n",
      "2800000 alignment record pairs processed.\n",
      "2900000 alignment record pairs processed.\n",
      "3000000 alignment record pairs processed.\n",
      "3089752 alignment record pairs processed.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "prefixes=(\"0M\" \"0P\" \"1a\" \"2P\" \"3L\" \"4C\")\n",
    "for pre in \"${prefixes[@]}\"; do\n",
    "  bam_files=()\n",
    "  for bam in sorted_bam/\"$pre\"*.bam; do\n",
    "    bam_files+=($bam)\n",
    "  done\n",
    "  echo counting ${bam_files[@]}\n",
    "    htseq-count --format=bam --order=name --stranded=yes \"${bam_files[@]}\" Homo_sapiens.GRCh38.113.gtf > htseqModResults/\"$pre\"_output.csv\n",
    "  echo output to htseqModResults/\"$pre\"_output.csv\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b43842a5-87bd-491f-a2db-abfa8716d24a",
   "metadata": {},
   "source": [
    "# DESeq 2\n",
    "This is used to get differentially expressed genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b8f5b68d-e472-4af0-835f-705269c68877",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "Bioconductor version 3.14 (BiocManager 1.30.25), R 4.1.2 (2021-11-01)\n",
      "Installing package(s) 'DESeq2'\n",
      "also installing the dependencies ‘Biostrings’, ‘curl’, ‘RCurl’, ‘KEGGREST’, ‘httr’, ‘GenomeInfoDb’, ‘AnnotationDbi’, ‘annotate’, ‘GenomicRanges’, ‘SummarizedExperiment’, ‘genefilter’, ‘geneplotter’\n",
      "\n",
      "trying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/Biostrings_2.62.0.tar.gz'\n",
      "Content type 'application/octet-stream' length 12645827 bytes (12.1 MB)\n",
      "==================================================\n",
      "downloaded 12.1 MB\n",
      "\n",
      "trying URL 'https://cloud.r-project.org/src/contrib/curl_6.0.1.tar.gz'\n",
      "Content type 'application/x-gzip' length 932716 bytes (910 KB)\n",
      "==================================================\n",
      "downloaded 910 KB\n",
      "\n",
      "trying URL 'https://cloud.r-project.org/src/contrib/RCurl_1.98-1.16.tar.gz'\n",
      "Content type 'application/x-gzip' length 728943 bytes (711 KB)\n",
      "==================================================\n",
      "downloaded 711 KB\n",
      "\n",
      "trying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/KEGGREST_1.34.0.tar.gz'\n",
      "Content type 'application/octet-stream' length 25778 bytes (25 KB)\n",
      "==================================================\n",
      "downloaded 25 KB\n",
      "\n",
      "trying URL 'https://cloud.r-project.org/src/contrib/httr_1.4.7.tar.gz'\n",
      "Content type 'application/x-gzip' length 118500 bytes (115 KB)\n",
      "==================================================\n",
      "downloaded 115 KB\n",
      "\n",
      "trying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/GenomeInfoDb_1.30.1.tar.gz'\n",
      "Content type 'application/octet-stream' length 3457147 bytes (3.3 MB)\n",
      "==================================================\n",
      "downloaded 3.3 MB\n",
      "\n",
      "trying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/AnnotationDbi_1.56.2.tar.gz'\n",
      "Content type 'application/octet-stream' length 4330396 bytes (4.1 MB)\n",
      "==================================================\n",
      "downloaded 4.1 MB\n",
      "\n",
      "trying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/annotate_1.72.0.tar.gz'\n",
      "Content type 'application/octet-stream' length 1723060 bytes (1.6 MB)\n",
      "==================================================\n",
      "downloaded 1.6 MB\n",
      "\n",
      "trying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/GenomicRanges_1.46.1.tar.gz'\n",
      "Content type 'application/octet-stream' length 1171939 bytes (1.1 MB)\n",
      "==================================================\n",
      "downloaded 1.1 MB\n",
      "\n",
      "trying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/SummarizedExperiment_1.24.0.tar.gz'\n",
      "Content type 'application/octet-stream' length 1524667 bytes (1.5 MB)\n",
      "==================================================\n",
      "downloaded 1.5 MB\n",
      "\n",
      "trying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/genefilter_1.76.0.tar.gz'\n",
      "Content type 'application/octet-stream' length 807639 bytes (788 KB)\n",
      "==================================================\n",
      "downloaded 788 KB\n",
      "\n",
      "trying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/geneplotter_1.72.0.tar.gz'\n",
      "Content type 'application/octet-stream' length 1437803 bytes (1.4 MB)\n",
      "==================================================\n",
      "downloaded 1.4 MB\n",
      "\n",
      "trying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/DESeq2_1.34.0.tar.gz'\n",
      "Content type 'application/octet-stream' length 2088431 bytes (2.0 MB)\n",
      "==================================================\n",
      "downloaded 2.0 MB\n",
      "\n",
      "* installing *source* package ‘curl’ ...\n",
      "** package ‘curl’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "Package libcurl was not found in the pkg-config search path.\n",
      "Perhaps you should add the directory containing `libcurl.pc'\n",
      "to the PKG_CONFIG_PATH environment variable\n",
      "No package 'libcurl' found\n",
      "Package libcurl was not found in the pkg-config search path.\n",
      "Perhaps you should add the directory containing `libcurl.pc'\n",
      "to the PKG_CONFIG_PATH environment variable\n",
      "No package 'libcurl' found\n",
      "Using PKG_CFLAGS=\n",
      "Using PKG_LIBS=-lcurl\n",
      "--------------------------- [ANTICONF] --------------------------------\n",
      "Configuration failed because libcurl was not found. Try installing:\n",
      " * deb: libcurl4-openssl-dev (Debian, Ubuntu, etc)\n",
      " * rpm: libcurl-devel (Fedora, CentOS, RHEL)\n",
      "If libcurl is already installed, check that 'pkg-config' is in your\n",
      "PATH and PKG_CONFIG_PATH contains a libcurl.pc file. If pkg-config\n",
      "is unavailable you can set INCLUDE_DIR and LIB_DIR manually via:\n",
      "R CMD INSTALL --configure-vars='INCLUDE_DIR=... LIB_DIR=...'\n",
      "-------------------------- [ERROR MESSAGE] ---------------------------\n",
      "<stdin>:1:10: fatal error: curl/curl.h: No such file or directory\n",
      "compilation terminated.\n",
      "--------------------------------------------------------------------\n",
      "ERROR: configuration failed for package ‘curl’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/curl’\n",
      "* installing *source* package ‘RCurl’ ...\n",
      "** package ‘RCurl’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "checking for curl-config... no\n",
      "Cannot find curl-config\n",
      "ERROR: configuration failed for package ‘RCurl’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/RCurl’\n",
      "ERROR: dependency ‘curl’ is not available for package ‘httr’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/httr’\n",
      "ERROR: dependency ‘RCurl’ is not available for package ‘GenomeInfoDb’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/GenomeInfoDb’\n",
      "ERROR: dependency ‘GenomeInfoDb’ is not available for package ‘Biostrings’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/Biostrings’\n",
      "ERROR: dependency ‘GenomeInfoDb’ is not available for package ‘GenomicRanges’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/GenomicRanges’\n",
      "ERROR: dependencies ‘httr’, ‘Biostrings’ are not available for package ‘KEGGREST’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/KEGGREST’\n",
      "ERROR: dependencies ‘GenomicRanges’, ‘GenomeInfoDb’ are not available for package ‘SummarizedExperiment’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/SummarizedExperiment’\n",
      "ERROR: dependency ‘KEGGREST’ is not available for package ‘AnnotationDbi’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/AnnotationDbi’\n",
      "ERROR: dependencies ‘AnnotationDbi’, ‘httr’ are not available for package ‘annotate’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/annotate’\n",
      "ERROR: dependencies ‘AnnotationDbi’, ‘annotate’ are not available for package ‘genefilter’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/genefilter’\n",
      "ERROR: dependencies ‘annotate’, ‘AnnotationDbi’ are not available for package ‘geneplotter’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/geneplotter’\n",
      "ERROR: dependencies ‘GenomicRanges’, ‘SummarizedExperiment’, ‘genefilter’, ‘geneplotter’ are not available for package ‘DESeq2’\n",
      "* removing ‘/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/DESeq2’\n",
      "\n",
      "The downloaded source packages are in\n",
      "\t‘/tmp/Rtmp6tlogr/downloaded_packages’\n",
      "Installation paths not writeable, unable to update packages\n",
      "  path: /usr/lib/R/library\n",
      "  packages:\n",
      "    boot, class, cluster, codetools, foreign, KernSmooth, lattice, mgcv, nlme,\n",
      "    nnet, rpart, spatial, survival\n",
      "Old packages: 'RcppArmadillo'\n",
      "There were 13 warnings (use warnings() to see them)\n",
      "Error in library(DESeq2) : there is no package called ‘DESeq2’\n",
      "Execution halted\n"
     ]
    }
   ],
   "source": [
    "!Rscript deseq2.R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99ff3c04-7ce3-4a5c-8555-1a3d405ad3c8",
   "metadata": {},
   "source": [
    "# ChEA3\n",
    "This script uses the ChEA3 API to perform transcription factor network analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d0c78c66-6ce4-47b8-8656-ab7775556414",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished writing results to chea3_output.csv\n"
     ]
    }
   ],
   "source": [
    "# Run ChEA3 on each dataframe\n",
    "\n",
    "import csv\n",
    "import requests\n",
    "import pandas as pd\n",
    "import sanbomics.tools\n",
    "\n",
    "MATRIGEL = \"00Matrigel\"\n",
    "PLL = \"00PLL\"\n",
    "AGRIN = \"01Agrin\"\n",
    "PAN_LAMANIN = \"02Pan-lamanin\"\n",
    "LAMANIN211 = \"03Lamanin211\"\n",
    "COLLAGENI = \"04Collagen_I\"\n",
    "\n",
    "# load the gene names from a CSV file.\n",
    "def get_gene_set(filename) -> list:\n",
    "    with open(filename, \"r\") as f:\n",
    "        data = csv.reader(f)\n",
    "\n",
    "        # return item in first column of each row (gene name)\n",
    "        return list(line[0] for line in data)\n",
    "\n",
    "# converts names like ENSG00000188157 to AGRN\n",
    "def map_genes(ensg_names: list) -> list:\n",
    "    mapper = sanbomics.tools.id_map(\"human\").mapper\n",
    "    return list(mapper[gene_name] for gene_name in ensg_names)\n",
    "\n",
    "# Creates a CSV file with ChEA3 using the given differentailly expressed\n",
    "# gene set and output filename.\n",
    "def make_chea3_tsv(gene_set, output_filename):\n",
    "    # The ChEA3 API URL.\n",
    "    url = \"https://maayanlab.cloud/chea3/api/enrich/\"\n",
    "\n",
    "    # Sends a JSON query to the ChEA3 API.\n",
    "    # https://maayanlab.cloud/chea3/\n",
    "    query_data = {\n",
    "        \"query_name\":\"myQuery\",\n",
    "        \"gene_set\":gene_set\n",
    "    }\n",
    "    result = requests.post(url, json=query_data)\n",
    "    assert result.ok, f\"Request failed with error code: {result.status_code}.\\\n",
    "    Check the ChEA3 website to make sure the request was correct.\"\n",
    "\n",
    "    # ChEA3 produces several tables (objects) in its JSON response.\n",
    "    # we get items from the table with this name:\n",
    "    entry = \"ENCODE--ChIP-seq\"\n",
    "    tfs = result.json()[entry]\n",
    "\n",
    "    with open(output_filename, \"w\") as tsvfile:\n",
    "        # Each item in the tfs is a dict. We make the table header using its keys.\n",
    "        fieldnames = tfs[0].keys()\n",
    "\n",
    "        writer = csv.DictWriter(tsvfile, fieldnames, delimiter=\"\\t\")\n",
    "        writer.writeheader()\n",
    "        for row in tfs:\n",
    "            writer.writerow(row)\n",
    "\n",
    "    print(f\"Finished writing results to {output_filename}\")\n",
    "\n",
    "# maps (control, treatment) to dataframes\n",
    "frames = {\n",
    "    (MATRIGEL, AGRIN): oneA_0Msigs,\n",
    "    (MATRIGEL, PAN_LAMANIN): twoP_0Msigs,\n",
    "    (MATRIGEL, LAMANIN211): threeL_0Msigs,\n",
    "    (MATRIGEL, COLLAGENI): fourC_0Msigs,\n",
    "    (PLL, AGRIN): oneA_0Psigs,\n",
    "    (PLL, PAN_LAMANIN): twoP_0Psigs,\n",
    "    (PLL, LAMANIN211): threeL_0Psigs,\n",
    "    (PLL, COLLAGENI): fourC_0Psigs\n",
    "}\n",
    "\n",
    "proteins = list(frames.keys()) + list(frames.values())\n",
    "for protein in proteins:\n",
    "    assert \" \" not in protein, f\"Protein with name '{protein}' should not have a space\"\n",
    "\n",
    "for (pair, frame) in frames.items():\n",
    "    control, treatment = pair\n",
    "\n",
    "    print(f\"Generating TSV for {control}, {treatment}...\")\n",
    "    gene_set = list(symbol for symbol in frame[\"Symbol\"] if isinstance(symbol, str))\n",
    "\n",
    "    make_chea3_tsv(gene_set, f\"chea3_results/{control}_vs_{treatment}.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "81c74759-106a-40e2-a41f-a75c8d4f0d0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print the result's data (warning: long)\n",
    "# result.json()\n",
    "\n",
    "# Print the result's text (JSON if successful, HTML if failure)\n",
    "# result.text"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a82c9b46-23f2-476f-bc66-e381e91dc599",
   "metadata": {},
   "source": [
    "# Step 6: Graph of transcription factors\n",
    "Generates graphs for each control-treatment pair using an R program written by Randall J. Eck."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a62a9a1b-30c2-42d0-af8e-21654bf23ad8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ahdog/Projects/csci597k_final_project\n",
      "\u001b[?25h\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hsample_ENCODE_ChIP-seq.tsv sample_ENCODE_ChIP-seq.png\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"Reading from sample_ENCODE_ChIP-seq.tsv printing to sample_ENCODE_ChIP-seq.png\"\n",
      "\u001b[?25h\u001b[1mRows: \u001b[22m\u001b[34m118\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m12\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): Query Name, Set_name, TF, Library, Overlapping_Genes\n",
      "\u001b[32mdbl\u001b[39m (7): Rank, Scaled Rank, Intersect, Set length, FET p-value, FDR, Odds Ratio\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hWarning message:\n",
      "\u001b[1m\u001b[22mRemoved 1 row containing missing values or values outside the scale range\n",
      "(`geom_point()`). \n",
      "\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "!pwd\n",
    "\n",
    "!Rscript ./transcriptionFactors.R sample_ENCODE_ChIP-seq.tsv sample_ENCODE_ChIP-seq.png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42f21350-c8b2-46a2-bdd8-4e02fa1f3466",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
