{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "45821872-0790-462c-b2cd-124e786bb4f0",
   "metadata": {},
   "source": [
    "# RNA STAR\n",
    "Here we perform sequence alignments on the input FASTQ files. This produces `.sam` files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e1cda9f-0f79-4d65-a85f-280d8d14b007",
   "metadata": {},
   "outputs": [],
   "source": [
    "!apt install rna-star"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d005624-710b-44b3-b79a-25a9069170f8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f32f7068-d6d3-496f-854e-3be01da270af",
   "metadata": {},
   "source": [
    "# sambamba\n",
    "Here we convert the `.sam` files to `.bam` files and sort them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "443e22f6-7b33-4d29-a86a-fc3fa445ea23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading package lists... Done\n",
      "Building dependency tree... Done\n",
      "Reading state information... Done\n",
      "The following additional packages will be installed:\n",
      "  libllvm11 libphobos2-ldc-shared98\n",
      "The following NEW packages will be installed:\n",
      "  libllvm11 libphobos2-ldc-shared98 sambamba\n",
      "0 upgraded, 3 newly installed, 0 to remove and 1 not upgraded.\n",
      "Need to get 21.8 MB of archives.\n",
      "After this operation, 92.9 MB of additional disk space will be used.\n",
      "Get:1 http://mirrors.ocf.berkeley.edu/ubuntu jammy/universe amd64 libllvm11 amd64 1:11.1.0-6 [19.6 MB]\n",
      "Get:2 http://mirrors.ocf.berkeley.edu/ubuntu jammy/universe amd64 libphobos2-ldc-shared98 amd64 1:1.28.0-1ubuntu1 [1,487 kB]\n",
      "Get:3 http://mirrors.ocf.berkeley.edu/ubuntu jammy/universe amd64 sambamba amd64 0.8.2+dfsg-2 [695 kB]\n",
      "Fetched 21.8 MB in 1s (21.4 MB/s)m\u001b[33m\n",
      "\n",
      "\u001b7\u001b[0;23r\u001b8\u001b[1ASelecting previously unselected package libllvm11:amd64.\n",
      "(Reading database ... 414490 files and directories currently installed.)Reading database ... 65%(Reading database ... 70%(Reading database ... 75%(Reading database ... 80%(Reading database ... 85%(Reading database ... 90%(Reading database ... 95%(Reading database ... 100%\n",
      "Preparing to unpack .../libllvm11_1%3a11.1.0-6_amd64.deb ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [  0%]\u001b[49m\u001b[39m [..........................................................] \u001b8\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [  8%]\u001b[49m\u001b[39m [####......................................................] \u001b8Unpacking libllvm11:amd64 (1:11.1.0-6) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 15%]\u001b[49m\u001b[39m [########..................................................] \u001b8Selecting previously unselected package libphobos2-ldc-shared98:amd64.\n",
      "Preparing to unpack .../libphobos2-ldc-shared98_1%3a1.28.0-1ubuntu1_amd64.deb ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 23%]\u001b[49m\u001b[39m [#############.............................................] \u001b8Unpacking libphobos2-ldc-shared98:amd64 (1:1.28.0-1ubuntu1) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 31%]\u001b[49m\u001b[39m [#################.........................................] \u001b8Selecting previously unselected package sambamba.\n",
      "Preparing to unpack .../sambamba_0.8.2+dfsg-2_amd64.deb ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 38%]\u001b[49m\u001b[39m [######################....................................] \u001b8Unpacking sambamba (0.8.2+dfsg-2) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 46%]\u001b[49m\u001b[39m [##########################................................] \u001b8Setting up libllvm11:amd64 (1:11.1.0-6) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 54%]\u001b[49m\u001b[39m [###############################...........................] \u001b8\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 62%]\u001b[49m\u001b[39m [###################################.......................] \u001b8Setting up libphobos2-ldc-shared98:amd64 (1:1.28.0-1ubuntu1) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 69%]\u001b[49m\u001b[39m [########################################..................] \u001b8\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 77%]\u001b[49m\u001b[39m [############################################..............] \u001b8Setting up sambamba (0.8.2+dfsg-2) ...\n",
      "\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 85%]\u001b[49m\u001b[39m [#################################################.........] \u001b8\u001b7\u001b[24;0f\u001b[42m\u001b[30mProgress: [ 92%]\u001b[49m\u001b[39m [#####################################################.....] \u001b8Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
      "Processing triggers for man-db (2.10.2-1) ...\n",
      "\n",
      "\u001b7\u001b[0;24r\u001b8\u001b[1A\u001b[J"
     ]
    }
   ],
   "source": [
    "!# password.txt is a text file with your password.\n",
    "!# If this gives an error (probably because you don't have sudo permissions),\n",
    "!# try running apt install -y sambamba instead.\n",
    "!sudo -S apt install -y sambamba < password.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b017d401-4698-4d93-92b4-bb6f2dd47e3e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dir results/00MatrigelCoating/\n",
      "dir results/00PLLCoating/\n",
      "dir results/01agrin/\n",
      "dir results/02PanLaminin/\n",
      "dir results/03Laminin211/\n",
      "dir results/04CollagenI/\n",
      "dir results/_STARtmp/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# note that bash scripts don't make output in Jupyter notebooks until they finish (sad!)\n",
    "\n",
    "mkdir -p ./sorted_bam\n",
    "for dir in results/*/; do\n",
    "  echo dir $dir\n",
    "  if [ -d \"$dir\" ]; then\n",
    "    for subdir in \"$dir\"/*/; do\n",
    "      subdir_name=$(basename \"$subdir\")\n",
    "      if [ -f \"$subdir/Aligned.out.sam\" ]; then\n",
    "        sam_file=\"$subdir/Aligned.out.sam\"\n",
    "        group=\"${dir:9:3}\"\n",
    "        short=\"${subdir_name:0:12}\"\n",
    "        output=$group-$short\n",
    "        sambamba view -S --format=bam $sam_file | sambamba sort -n -o sorted_bam/${output}.bam /dev/stdin\n",
    "      fi\n",
    "    done\n",
    "  fi\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f683af31-b47d-4261-95d0-d20d54025711",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ran sambamba on results/01agrin/lib103916_L2-ds.d1b626e6358749c5aa8c7a900453bf55/results/01agrin/lib103916_L2-ds.d1b626e6358749c5aa8c7a900453bf55/Aligned.out.sam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ran sambamba on results/01agrin/lib103917_L1-ds.6cb7a56486a743aa97aef4b4492c71db/results/01agrin/lib103917_L1-ds.6cb7a56486a743aa97aef4b4492c71db/Aligned.out.sam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ran sambamba on results/01agrin/lib103916_L1-ds.cf006f61b73c426b81c8530ea73ed130/results/01agrin/lib103916_L1-ds.cf006f61b73c426b81c8530ea73ed130/Aligned.out.sam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n",
      "\n",
      "sambamba 0.8.2\n",
      " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
      "    LDC 1.28.0 / DMD v2.098.0 / LLVM13.0.0 / bootstrap LDC - the LLVM D compiler (1.28.0)\n",
      "\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_26428/558986621.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     17\u001b[0m             \u001b[0moutput\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf\"{group}-{short}\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m             \u001b[0;31m# run sambamba on the file and put it in sorted_bam\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m             \u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"sambamba view -S --format=bam {sam_file} | sambamba sort -n -o sorted_bam/{output}.bam /dev/stdin\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mshell\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcheck\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Ran sambamba on {subfolderpath}/{sam_file}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.10/subprocess.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    503\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0mPopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mpopenargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    504\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 505\u001b[0;31m             \u001b[0mstdout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommunicate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    506\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mTimeoutExpired\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    507\u001b[0m             \u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkill\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.10/subprocess.py\u001b[0m in \u001b[0;36mcommunicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1144\u001b[0m                 \u001b[0mstderr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1145\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1146\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1147\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1148\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.10/subprocess.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1207\u001b[0m             \u001b[0mendtime\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_time\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1208\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1209\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_wait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1210\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1211\u001b[0m             \u001b[0;31m# https://bugs.python.org/issue25942\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.10/subprocess.py\u001b[0m in \u001b[0;36m_wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1957\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1958\u001b[0m                             \u001b[0;32mbreak\u001b[0m  \u001b[0;31m# Another thread waited.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1959\u001b[0;31m                         \u001b[0;34m(\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msts\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_try_wait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1960\u001b[0m                         \u001b[0;31m# Check the pid and loop as waitpid has been known to\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1961\u001b[0m                         \u001b[0;31m# return 0 even without WNOHANG in odd situations.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.10/subprocess.py\u001b[0m in \u001b[0;36m_try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1915\u001b[0m             \u001b[0;34m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1916\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1917\u001b[0;31m                 \u001b[0;34m(\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msts\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwaitpid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwait_flags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1918\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mChildProcessError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1919\u001b[0m                 \u001b[0;31m# This happens if SIGCLD is set to be ignored or waiting\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import pathlib\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "pathlib.Path(\"sorted_bam\").mkdir(exist_ok=True)\n",
    "for folder in os.listdir(\"results\"):\n",
    "    folderpath = f\"results/{folder}\"\n",
    "    if not os.path.isdir(folderpath):\n",
    "        continue\n",
    "    \n",
    "    for subfolder in os.listdir(folderpath):\n",
    "        subfolderpath = f\"{folderpath}/{subfolder}\"\n",
    "        sam_file = f\"{subfolderpath}/Aligned.out.sam\"\n",
    "        if os.path.isfile(sam_file):\n",
    "            group = folder[9:12]\n",
    "            short = subfolder[0:13]\n",
    "            output = f\"{group}-{short}\"\n",
    "            # run sambamba on the file and put it in sorted_bam\n",
    "            subprocess.run(f\"sambamba view -S --format=bam {sam_file} | sambamba sort -n -o sorted_bam/{output}.bam /dev/stdin\", shell=True, check=True)\n",
    "            print(f\"Ran sambamba on {subfolderpath}/{sam_file}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d0d8c64-faac-4d28-8d94-3c469a12b556",
   "metadata": {},
   "source": [
    "# htseq-count\n",
    "Here we get sequence counts, and store them in CSV files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "35d9670c-54cb-4d2b-a26e-fa923bc6548d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sorting sorted_bam/1ag-lib103915_L1.bam ...\n",
      "sorting sorted_bam/1ag-lib103915_L2.bam ...\n",
      "sorting sorted_bam/1ag-lib103916_L1.bam ...\n",
      "sorting sorted_bam/1ag-lib103916_L2.bam ...\n",
      "sorting sorted_bam/1ag-lib103917_L1.bam ...\n",
      "sorting sorted_bam/1ag-lib103917_L2.bam ...\n",
      "output to htseqResults/1a_output.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2667906 alignment record pairs processed.\n",
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2700000 alignment record pairs processed.\n",
      "2800000 alignment record pairs processed.\n",
      "2900000 alignment record pairs processed.\n",
      "2933419 alignment record pairs processed.\n",
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2700000 alignment record pairs processed.\n",
      "2800000 alignment record pairs processed.\n",
      "2804493 alignment record pairs processed.\n",
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2700000 alignment record pairs processed.\n",
      "2800000 alignment record pairs processed.\n",
      "2876643 alignment record pairs processed.\n",
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2700000 alignment record pairs processed.\n",
      "2800000 alignment record pairs processed.\n",
      "2900000 alignment record pairs processed.\n",
      "2916191 alignment record pairs processed.\n",
      "100000 GFF lines processed.\n",
      "200000 GFF lines processed.\n",
      "300000 GFF lines processed.\n",
      "400000 GFF lines processed.\n",
      "500000 GFF lines processed.\n",
      "600000 GFF lines processed.\n",
      "700000 GFF lines processed.\n",
      "800000 GFF lines processed.\n",
      "900000 GFF lines processed.\n",
      "1000000 GFF lines processed.\n",
      "1100000 GFF lines processed.\n",
      "1200000 GFF lines processed.\n",
      "1300000 GFF lines processed.\n",
      "1400000 GFF lines processed.\n",
      "1500000 GFF lines processed.\n",
      "1600000 GFF lines processed.\n",
      "1700000 GFF lines processed.\n",
      "1800000 GFF lines processed.\n",
      "1900000 GFF lines processed.\n",
      "2000000 GFF lines processed.\n",
      "2100000 GFF lines processed.\n",
      "2200000 GFF lines processed.\n",
      "2300000 GFF lines processed.\n",
      "2400000 GFF lines processed.\n",
      "2500000 GFF lines processed.\n",
      "2600000 GFF lines processed.\n",
      "2700000 GFF lines processed.\n",
      "2800000 GFF lines processed.\n",
      "2900000 GFF lines processed.\n",
      "3000000 GFF lines processed.\n",
      "3100000 GFF lines processed.\n",
      "3200000 GFF lines processed.\n",
      "3300000 GFF lines processed.\n",
      "3400000 GFF lines processed.\n",
      "3500000 GFF lines processed.\n",
      "3600000 GFF lines processed.\n",
      "3700000 GFF lines processed.\n",
      "3800000 GFF lines processed.\n",
      "3900000 GFF lines processed.\n",
      "4000000 GFF lines processed.\n",
      "4100000 GFF lines processed.\n",
      "4114450 GFF lines processed.\n",
      "100000 alignment record pairs processed.\n",
      "200000 alignment record pairs processed.\n",
      "300000 alignment record pairs processed.\n",
      "400000 alignment record pairs processed.\n",
      "500000 alignment record pairs processed.\n",
      "600000 alignment record pairs processed.\n",
      "700000 alignment record pairs processed.\n",
      "800000 alignment record pairs processed.\n",
      "900000 alignment record pairs processed.\n",
      "1000000 alignment record pairs processed.\n",
      "1100000 alignment record pairs processed.\n",
      "1200000 alignment record pairs processed.\n",
      "1300000 alignment record pairs processed.\n",
      "1400000 alignment record pairs processed.\n",
      "1500000 alignment record pairs processed.\n",
      "1600000 alignment record pairs processed.\n",
      "1700000 alignment record pairs processed.\n",
      "1800000 alignment record pairs processed.\n",
      "1900000 alignment record pairs processed.\n",
      "2000000 alignment record pairs processed.\n",
      "2100000 alignment record pairs processed.\n",
      "2200000 alignment record pairs processed.\n",
      "2300000 alignment record pairs processed.\n",
      "2400000 alignment record pairs processed.\n",
      "2500000 alignment record pairs processed.\n",
      "2600000 alignment record pairs processed.\n",
      "2700000 alignment record pairs processed.\n",
      "2800000 alignment record pairs processed.\n",
      "2900000 alignment record pairs processed.\n",
      "3000000 alignment record pairs processed.\n",
      "3089752 alignment record pairs processed.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "prefixes=(\"0M\" \"0P\" \"1a\" \"2P\" \"3L\" \"4C\")\n",
    "for pre in \"${prefixes[@]}\"; do\n",
    "  bam_files=()\n",
    "  for bam in sorted_bam/\"$pre\"*.bam; do\n",
    "    bam_files+=($bam)\n",
    "  done\n",
    "  echo counting ${bam_files[@]}\n",
    "    htseq-count --format=bam --order=name --stranded=yes \"${bam_files[@]}\" Homo_sapiens.GRCh38.113.gtf > htseqModResults/\"$pre\"_output.csv\n",
    "  echo output to htseqModResults/\"$pre\"_output.csv\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b43842a5-87bd-491f-a2db-abfa8716d24a",
   "metadata": {},
   "source": [
    "# DESeq 2\n",
    "This is used to get differentially expressed genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "58f19bcf-210d-4b90-b7fa-6250cdd8775e",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 0
    },
    "id": "81qBMgjTZQHl",
    "outputId": "b051a466-3813-4e6a-d692-5c51cda64223"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "from sanbomics.tools import id_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ef77f9a2-18ee-46cc-8b41-e09ab9cdba88",
   "metadata": {
    "id": "BCjjTce1ZMDI"
   },
   "outputs": [],
   "source": [
    "def load_count_matrix():\n",
    "    \"\"\"\n",
    "    Load, merge, and preprocess count files into a single count matrix.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A preprocessed count matrix with gene IDs as columns and sample names as rows,\n",
    "                  ready for downstream analysis.\n",
    "    \"\"\"\n",
    "    counts_files = {\n",
    "        \"0M\": \"htseqResults/0M_output.csv\",\n",
    "        \"0P\": \"htseqResults/0P_output.csv\",\n",
    "        \"1A\": \"htseqResults/1a_output.csv\",\n",
    "        \"2P\": \"htseqResults/2P_output.csv\",\n",
    "        \"3L\": \"htseqResults/3L_output.csv\",\n",
    "        \"4C\": \"htseqResults/4C_output.csv\",\n",
    "    }\n",
    "\n",
    "    # Initialize an empty DataFrame\n",
    "    count_matrix = None\n",
    "\n",
    "    # Process each file\n",
    "    for sample, file_path in counts_files.items():\n",
    "        # Read the file with tab-separated values\n",
    "        df = pd.read_csv(file_path, sep=\"\\t\", header=None, names=[\"geneID\", sample])\n",
    "\n",
    "        # Merge all files into one count matrix\n",
    "        if count_matrix is None:\n",
    "            count_matrix = df\n",
    "        else:\n",
    "            count_matrix = pd.merge(count_matrix, df, on=\"geneID\", how=\"outer\")\n",
    "\n",
    "    # Set the geneID column as the index\n",
    "    count_matrix.set_index(\"geneID\", inplace=True)\n",
    "\n",
    "    # Replace NaN values with zeros (common for gene count data)\n",
    "    count_matrix.fillna(0, inplace=True)\n",
    "\n",
    "    # Filter out rows (genes) with zero counts across all samples\n",
    "    count_matrix = count_matrix[count_matrix.sum(axis=1) > 0]\n",
    "\n",
    "    count_matrix = count_matrix[count_matrix.index.str.startswith('E')]\n",
    "\n",
    "    # Transpose the count matrix to match DESeq2 format (samples as rows, genes as columns)\n",
    "    count_matrix = count_matrix.T\n",
    "\n",
    "    return count_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5bc9c505-423b-452a-a247-bddf573a8ce5",
   "metadata": {
    "id": "VZptgdjgXhNc"
   },
   "outputs": [],
   "source": [
    "def run_combined_test(\n",
    "    count_matrix, metadata, species='human'\n",
    "):\n",
    "    \"\"\"\n",
    "    Run DESeq2 differential expression test for the combined group (RS) against the control group.\n",
    "    Includes gene mapping using Sanbomics.\n",
    "\n",
    "    Parameters:\n",
    "    - count_matrix (pd.DataFrame): Transposed count matrix with samples as rows and genes as columns.\n",
    "    - control_label (str): The label for the control group in the metadata.\n",
    "    - combined_label (str): The label for the combined treatment group in the metadata.\n",
    "    - padj_thresh (float): Threshold for adjusted p-value to consider significance.\n",
    "    - lfc_thresh (float): Threshold for log2 fold change to consider significance.\n",
    "    - species (str): The species for the gene mapping (default is 'human').\n",
    "\n",
    "    Returns:\n",
    "    - pd.DataFrame: A DataFrame with significant results (with mapped symbols) for the combined test.\n",
    "    \"\"\"\n",
    "\n",
    "    # Initialize DeseqDataSet\n",
    "    dds = DeseqDataSet(counts=count_matrix, metadata=metadata, design_factors=\"Condition\")\n",
    "    dds.deseq2()\n",
    "\n",
    "    # Normalize total counts in the AnnData object\n",
    "    sc.pp.normalize_total(dds, target_sum=1e4)\n",
    "\n",
    "    sc.tl.pca(dds)\n",
    "\n",
    "    sc.pl.pca(dds, color = 'Condition', size = 200)\n",
    "\n",
    "    # Run differential expression test for the combined group\n",
    "    stat_res = DeseqStats(dds, contrast=('Condition', 'RS', 'C'))\n",
    "\n",
    "    # Run Wald tests\n",
    "    stat_res.summary()\n",
    "\n",
    "    # Return results\n",
    "    res = stat_res.results_df\n",
    "\n",
    "    # Map gene IDs to symbols\n",
    "    mapper = id_map(species=species)\n",
    "    res['Symbol'] = res.index.map(mapper.mapper)\n",
    "\n",
    "    return res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b8f79958-f2ae-4eb2-bb14-ab19c44c81e1",
   "metadata": {
    "id": "SEq9b9LnXvYK"
   },
   "outputs": [],
   "source": [
    "def run_individual_tests(\n",
    "    count_matrix, treatments, metadata, species='human'\n",
    "):\n",
    "    \"\"\"\n",
    "    Run DESeq2 differential expression tests for individual treatments against the control group.\n",
    "    Includes gene mapping using Sanbomics.\n",
    "\n",
    "    Parameters:\n",
    "    - count_matrix (pd.DataFrame): Transposed count matrix with samples as rows and genes as columns.\n",
    "    - treatments (list): List of treatment labels to test against the control.\n",
    "    - control_label (str): The label for the control group in the metadata.\n",
    "    - padj_thresh (float): Threshold for adjusted p-value to consider significance.\n",
    "    - lfc_thresh (float): Threshold for log2 fold change to consider significance.\n",
    "    - species (str): The species for the gene mapping (default is 'human').\n",
    "\n",
    "    Returns:\n",
    "    - dict: A dictionary with treatment labels as keys and significant results (with mapped symbols) as DataFrames.\n",
    "    \"\"\"\n",
    "\n",
    "    # Initialize DeseqDataSet\n",
    "    dds = DeseqDataSet(counts=count_matrix, metadata=metadata, design_factors=\"Condition\")\n",
    "    dds.deseq2()\n",
    "\n",
    "    # Normalize total counts in the AnnData object\n",
    "    sc.pp.normalize_total(dds, target_sum=1e4)\n",
    "\n",
    "    sc.tl.pca(dds)\n",
    "\n",
    "    sc.pl.pca(dds, color = 'Condition', size = 200)\n",
    "\n",
    "    results = {}  # Store results for each treatment\n",
    "    mapper = id_map(species=species)  # Initialize gene ID mapper\n",
    "\n",
    "    for treatment in treatments:\n",
    "        # Run differential expression test for each treatment vs control\n",
    "        stat_res = DeseqStats(dds, contrast=('Condition', treatment, 'C'))\n",
    "\n",
    "        # Run Wald tests\n",
    "        stat_res.summary()\n",
    "\n",
    "        # Return results\n",
    "        res = stat_res.results_df\n",
    "\n",
    "        # Map gene IDs to symbols\n",
    "        res['Symbol'] = res.index.map(mapper.mapper)\n",
    "\n",
    "        # Save the results to the dictionary\n",
    "        results[treatment] = res\n",
    "\n",
    "    return results\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7650502b-f561-4cd4-ba2a-81532307c060",
   "metadata": {
    "id": "PFuZVGwdZaY5"
   },
   "source": [
    "## Load and observe data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11e372a9-4f5c-42b2-861b-f6d2bd5b57df",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "0cflDlMOYV2M",
    "outputId": "dc1842d8-6023-4ad6-ebb8-07808f6c1f71"
   },
   "outputs": [],
   "source": [
    "# Load the count matrix\n",
    "count_matrix = load_count_matrix()\n",
    "\n",
    "# Display the first few rows of the processed count matrix\n",
    "print(count_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "c2f46f5a-a9bd-4f58-81cb-e4ef8065e82c",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "sJsz3SElafyU",
    "outputId": "2adac172-24e0-4a4a-ca0c-152ba054bc75"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Condition\n",
      "Sample          \n",
      "0M             C\n",
      "0P             C\n",
      "1A            RS\n",
      "2P            RS\n",
      "3L            RS\n",
      "4C            RS\n"
     ]
    }
   ],
   "source": [
    "# create a metadata df for samples and conditions\n",
    "metadata = pd.DataFrame(zip(count_matrix.index, ['C', 'C', 'RS', 'RS', 'RS', 'RS']), columns = ['Sample', 'Condition'])\n",
    "\n",
    "metadata = metadata.set_index('Sample')\n",
    "print(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "65644751-6105-41e9-bb0e-22bd29e7e23b",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "tgTP9mq1apo3",
    "outputId": "f52f402f-e8e2-4e71-86cd-bd45563baab3"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Condition\n",
      "Sample          \n",
      "0M             C\n",
      "0P             C\n",
      "1A            1A\n",
      "2P            2P\n",
      "3L            3L\n",
      "4C            4C\n"
     ]
    }
   ],
   "source": [
    "# create a metadata df for samples and conditions\n",
    "indTreatMetadata = pd.DataFrame(zip(count_matrix.index, ['C', 'C', '1A', '2P', '3L', '4C']), columns = ['Sample', 'Condition'])\n",
    "\n",
    "indTreatMetadata = indTreatMetadata.set_index('Sample')\n",
    "print(indTreatMetadata.head(6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b1e3605-8e73-4961-bb1d-1eb1e14ffbcb",
   "metadata": {
    "id": "yZQw1PWmX6nj"
   },
   "source": [
    "## Run all treatments against all controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "a1dfc8b0-456b-4d07-8301-fb2a1618eb4a",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "XYlySpaIfZFX",
    "outputId": "1b7c16f3-1988-413b-a0f7-a8c90f2fcc6f"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 29.51 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 1.38 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 31.48 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 25.21 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAmIAAAGvCAYAAAD4yLkRAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAeeElEQVR4nO3de5iVdb338c/ijMCAWG0YxOGkqCC4rwpLdxE+amQet6e0EjxmkVvKdKe1S9PtY9reeczCPJHUrh4T0Uy9Ss1t5qEUTcxCQzyAaCkHUeQw6/mDDTkbgRmZ4TczvF7/dK3f/Vvr/g5dl/O+1rrnXpVqtVoNAACbXYfSAwAAbKmEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGNCsrr322lQqlTzzzDNr1z7ykY/kIx/5SKOeP3HixAwaNKhFZgNobYQYtANPP/10PvOZz2TIkCHp1q1bampqsscee+Tiiy/OG2+8UXq8dcybNy9nnXVWZs6cWXoUgKI6lR4A2DQ///nPc9hhh6Vr1645+uijM3LkyCxfvjz33ntvTjvttMyaNStTpkwpOuMdd9zR4PG8efNy9tlnZ9CgQdl1110bHLvyyitTX1+/GacDKEeIQRs2Z86cfOITn0hdXV3uvPPO9O/ff+2xSZMm5amnnsrPf/7zghOu1qVLl0bv7dy5cwtOAtC6+GgS2rALLrggr732Wq666qoGEbbGsGHDcsoppyRJVq5cmXPOOSdDhw5N165dM2jQoJx55pl58803Gzxn0KBB2W+//XLvvfdmzJgx6datW4YMGZKpU6eu8/qzZs3Knnvume7du2fbbbfNueee+7bvZr31GrG7774773//+5MkxxxzTCqVSiqVSq699tokb3+N2NKlS3Pqqadm4MCB6dq1a4YPH55vfetbqVarDfZVKpV8/vOfz/Tp0zNy5Mh07do1I0aMyG233daof0+Aza1S/d//JQPajG233TZdu3bN008/vdG9EydOzHXXXZdDDz0048aNywMPPJCpU6fmoIMOyo033rh236BBg9KtW7csXLgwxx13XGpra3P11VfnkUceyR/+8IeMGDEiSfLiiy9m1KhRWblyZU455ZT06NEjU6ZMSffu3fPYY49lzpw5a4PqrRG2YMGCTJkyJV/72tdy4okn5kMf+lCSZPfdd8+QIUMyceLE3H333Wsv9q9Wq9lrr71y11135bjjjsuuu+6a22+/PTfffHMmT56cb3/722tnr1QqGT16dF566aV87nOfS69evXLJJZfkxRdfzLPPPpttttmmGf7VAZpRFWiTFi1aVE1SPfDAAze6d+bMmdUk1eOPP77B+pe+9KVqkuqdd965dq2urq6apHrPPfesXXvppZeqXbt2rZ566qlr1yZPnlxNUn3ggQca7Ovdu3c1SXXOnDlr18eOHVsdO3bs2scPPfRQNUn1mmuuWWfWCRMmVOvq6tY+nj59ejVJ9dxzz22w79BDD61WKpXqU089tXYtSbVLly4N1h599NFqkuqll166/n8ggEJ8NAlt1OLFi5MkvXr12ujeW2+9NUnyxS9+scH6qaeemiTrXEe28847r32nKkne/e53Z/jw4fnLX/7S4DU/8IEPZMyYMQ32ffKTn2ziT7Lx2Tt27Jh/+Zd/WWf2arWaX/ziFw3W99prrwwdOnTt41GjRqWmpqbB7ACthRCDNqqmpiZJsmTJko3unTt3bjp06JBhw4Y1WO/Xr1/69OmTuXPnNljfbrvt1nmNrbfeOq+++mqD19x+++3X2Td8+PBGzd9Yc+fOTW1t7TrBudNOO609/laNmR2gtRBi0EbV1NSktrY2jz/+eKOfU6lUGrWvY8eOb7tebQOXlLbl2YEtjxCDNmy//fbL008/nd/+9rcb3FdXV5f6+vrMnj27wfqCBQuycOHC1NXVNfncdXV167xekvzpT3/a6HMbG4RrzjNv3rx13vl78skn1x4HaKuEGLRhp59+enr06JHjjz8+CxYsWOf4008/nYsvvjj77rtvkuSiiy5qcPw///M/kyQf//jHm3zufffdN/fff38efPDBtWsvv/xypk2bttHn9ujRI0mycOHCRp1n1apVueyyyxqsf/vb306lUsnHPvaxpg0O0Iq4oSu0YUOHDs0Pf/jDHHHEEdlpp50a3Fn/vvvuy09/+tNMnDgxp5xySiZMmJApU6Zk4cKFGTt2bB588MFcd911OeiggzJu3Lgmn/v000/PD37wg4wfP77B7Svq6ury2GOPbXTuPn365Lvf/W569eqVHj16ZLfddsvgwYPX2bv//vtn3Lhx+cpXvpJnnnkmo0ePzh133JGbbropkydPbnBhPkBbI8SgjTvggAPy2GOP5cILL8xNN92UK664Il27ds2oUaPyH//xHznhhBOSJN///vczZMiQXHvttbnxxhvTr1+/nHHGGfn617/+js7bv3//3HXXXTn55JNz/vnnZ5tttslJJ52U2traHHfccRt8bufOnXPdddfljDPOyEknnZSVK1fmmmuuedsQ69ChQ2bMmJGvfe1r+fGPf5xrrrkmgwYNyoUXXrj2rz4B2io3dAUAKMQ1YgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKKTIfcTq6+szb9689OrVq0lfdQIAlFOtVrNkyZLU1tamQwfv5TSHIiE2b968DBw4sMSpAYBN9Nxzz2XbbbctPUa7UCTEevXqlWT1/5E1NTUlRgAAmmjx4sUZOHDg2t/jbLoiIbbm48iamhohBgBtjMuKmo8PeAEAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABRS5IauW5RqNZn/aLL8taTvkKSmtvREAEArIcRa0swfJvd8K3nl6dWPKx2THT6a7HNuss3QsrMBAMX5aLKl/ObiZPpn/x5hSVJdlfzp1uTqjyav/KXcbABAqyDEWsJrLyd3nrv+40tfTn51zuabBwBolYRYS3j0R8mq5Rve88ebk9df2TzzAACtkhBrCY352LF+RbLo+ZafBQBotYRYS+jep3H7uvVu0TEAgNZNiLWEkYdsfM+2Y5Kt61p+FgCg1RJiLaHfLsmO+63/eKVDMvZfN988AECrJMRayj9fmYw4OEml4XrX3snBU5Lt9yoyFgDQeriha0vpslVy2LXJuKeSJ6Ynby5J3rXD6jjrslXp6QCAVkCItbR3DUs+/KXSUwAArZCPJgEAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABTSqfQA7db8R5Olf01qBiTv2bH0NABAKyTEmtuTtyZ3npO89MTf1wa8L9nnnKRu93JzAQCtjo8mm9PjNyQ//mTDCEuSF36XTD0omfPfRcYCAFonIdZcVq1IfvHlpFq/nuNvJredsXlnAgBaNSHWXP58W7L0pQ3vWfCH5IXfb555AIBWT4g1l1fnNu8+AKDdE2LNpfvWzbsPAGj3hFhz2fHjSeetNrynV20y6EObZx4AoNUTYs2le5/kg5M2vGfs6UlHdwwBAFZTBc1p3FdW//Xk/d9JVi3/+3rnHsm4M5P3HVNuNgCg1RFizalSSfY+O9n95OTxnyVLX056D0hG/HPSrab0dABAKyPEWkKPdyW7nVh6CgCglXONGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUEin0gO0GYueT/54S7J8SfKuHZLh+yYdO5eeCgBow4TYxqx8M/n5F5OZP0qqq/6+3rNfsv/FyfDx5WYDANo0H01uzI0nJY9c3zDCkuS1F5Mffyp55jdl5gIA2jwhtiELnkhm/Wz9x+tXJHf/3803DwDQrgixDfnDTza+55n/ThbPa/lZAIB2R4htyOt/a9y+N15t2TkAgHZJiG1I7+02vqdDp6RX/5afBQBod4TYhux6VFLpuOE9O+6XbNV388wDALQrQmxDeg9IPvTF9R/vvnWy51c33zwAQLsixDZmz68m47+57sePg8cmx9yWvGv7MnMBAG2eG7o2xgdOSt5/fPLc/cmbr62Or22Glp4KAGjjhFhjdeyUDPqn0lMAAO2IjyYBAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQBbvBdffDEnn3xyhgwZkq5du2bgwIHZf//986tf/apFz9upRV8dAKAJVtVX8+CcV/LSkmV5T69uGTO4bzp2qLToOZ955pnsscce6dOnTy688MLssssuWbFiRW6//fZMmjQpTz75ZIudW4gBAK3CbY/Pz9k3P5H5i5atXevfu1u+vv/OGT+yf4ud93Of+1wqlUoefPDB9OjRY+36iBEjcuyxx7bYeRMfTQIArcBtj8/PZ69/uEGEJcmLi5bls9c/nNsen98i533llVdy2223ZdKkSQ0ibI0+ffq0yHnXEGIAQFGr6qs5++YnUn2bY2vWzr75iayqf7sdm+app55KtVrNjjvu2Oyv3RhCDAAo6sE5r6zzTthbVZPMX7QsD855pdnPXa02f9w1hRADAIp6acn6I+yd7GuK7bffPpVKpUUvyN+QJofY/Pnzc/311+fWW2/N8uXLGxxbunRpvvGNbzTbcABA+/eeXt2adV9T9O3bNx/96Edz+eWXZ+nSpescX7hwYbOf862aFGIPPfRQdt5550yaNCmHHnpoRowYkVmzZq09/tprr+Xss89u9iEBgPZrzOC+6d+7W9Z3k4pKVv/15JjBfVvk/JdffnlWrVqVMWPG5IYbbsjs2bPzxz/+MZdcckk++MEPtsg512hSiJ155pk5+OCD8+qrr2bBggXZe++9M3bs2DzyyCMtNR8A0M517FDJ1/ffOUnWibE1j7++/84tdj+xIUOG5OGHH864ceNy6qmnZuTIkdl7773zq1/9KldccUWLnHONSrUJV6n17ds3999/f3bYYYe1a+eff34uuOCC3H777dluu+1SW1ubVatWbfB1Fi9enN69e2fRokWpqal559MDAJtNS//+LnUfsZKafEPXZcsaXij35S9/OZ06dco+++yTq6++utkGAwC2LONH9s/eO/fb7HfWL6lJITZy5Mjcd999GTVqVIP1L33pS6mvr8+RRx7ZrMMBAFuWjh0q+eDQbUqPsdk06Rqxo48+Ovfee+/bHjv99NNz9tlnZ7vttmuWwQAA2rsmXSPWXFwjBgBtj9/fza9J74gtW7YsM2bMyJIlS9Y5tnjx4syYMSNvvvlmsw0HANCeNSnEvve97+Xiiy9Or1691jlWU1OTSy65JFdeeWWzDQcA0J41KcSmTZuWyZMnr/f45MmTM3Xq1E2dCQBgi9CkEJs9e3ZGjx693uOjRo3K7NmzN3koAIAtQZNCbOXKlXn55ZfXe/zll1/OypUrN3koAIAtQZNCbMSIEfnlL3+53uN33HFHRowYsclDAQBsCZoUYscee2zOOeec3HLLLescu/nmm/Pv//7vOfbYY5ttOABgC1O/Kpnz38kf/t/q/63f8NcmbqqJEyemUqmkUqmkc+fOGTx4cE4//fQG3yT061//OnvuuWf69u2brbbaKttvv30mTJiQ5cuXb/L5m3Rn/RNPPDH33HNPDjjggOy4444ZPnx4kuTJJ5/Mn//85xx++OE58cQTN3koAGAL9MSM5LZ/TRbP+/taTW0y/pvJzge02GnHjx+fa665JitWrMjvf//7TJgwIZVKJd/85jfzxBNPZPz48Tn55JNzySWXpHv37pk9e3ZuuOGGjX63dmO8oxu6/vSnP820adMye/bsVKvV7LDDDjnqqKNy+OGHN+r5bggHAG1Pi/7+fmJG8pOjk/zvLPmf75k8fGqLxNjEiROzcOHCTJ8+fe3aIYcckjlz5uThhx/ORRddlIsvvjhz5sxp9nMnTXxHbNWqVfnWt76VGTNmZPny5dlvv/1y1llnpXv37i0yHACwBahftfqdsHUiLP+zVklu+3Ky48eTDh1bdJTHH3889913X+rq6pIk/fr1y/z583PPPffkwx/+cLOfr0nXiJ133nk588wz07NnzwwYMCCXXHJJJk2a1OxDAQBbkLn3Nfw4ch3VZPELq/e1gFtuuSU9e/ZMt27dsssuu+Sll17KaaedliQ57LDDcuSRR2bs2LHp379/Dj744Fx22WVZvHhxs5y7SSE2derUfOc738ntt9+e6dOn5+abb860adNSX1/fLMMAAFug1xY0774mGjduXGbOnJkHHnggEyZMyDHHHJNDDjkkSdKxY8dcc801ef7553PBBRdkwIABOe+88zJixIjMnz9/k8/dpBB79tlns++++659vNdee6VSqWTevA1VLADABvT8h+bd10Q9evTIsGHDMnr06Fx99dV54IEHctVVVzXYM2DAgHz605/OZZddllmzZmXZsmX57ne/u8nnbvINXbt169ZgrXPnzlmxYsUmDwIAbKHqdl/915FrLsxfRyWpGbB6Xwvr0KFDzjzzzHz1q1/NG2+88bZ7tt566/Tv3z9Lly7d5PM16WL9arWaiRMnpmvXrmvXli1blpNOOik9evRYu/azn/1skwcDALYQHTquvkXFT47O6hh760X7/xNn489v8Qv11zjssMNy2mmn5fLLL0+vXr0yc+bMHHzwwRk6dGiWLVuWqVOnZtasWbn00ks3+VxNCrEJEyass/apT31qk4cAALZwOx+w+hYVb3sfsfNb9D5i/1unTp3y+c9/PhdccEFuvPHG3HvvvTnppJMyb9689OzZMyNGjMj06dMzduzYTT7XO7qP2KZyHzEAaHs2y+/v+lWr/zrytQWrrwmr232zvRNWQpPeEQMAaFEdOiaDP1R6is2mSRfrAwDQfLwjBgDtwYo3ksdvSJ65d/Xjut2TXQ5LOvv2m9ZMiAFAW/fcQ8l/HZksffnva4/+KPnlWcknfphs94Fio7FhPpoEgLZs8bxk2iENI2yN1/+WTDs8Wfjc5p+LRhFiANCWPXRVsmzR+o+/uSh56Pubbx6aRIgBQFv2x5sbsWdGy8/BOyLEAKAtW/H2X8PTwPLXW34O3hEhBgBt2Xt2bMSenVp+Dt4RIQYAbdn7jm2ePRQhxACgLdthfDLqiPUfH3lIstP+m28emsR9xACgLatUkoO+m9T+Y/LA95JX56xe33pQ8v4Tkg98bvUeWiUhBgBtXYcOyQc+m+x2UrLo+STVpPdAAdYGCDEAaC8qlaTPwNJT0ASuEQMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgkE6lB2hujz2/ML956m+pr1YzZnDfvH9Q39IjAQC8rXYTYvMXvZHP//CR/H7uqw3Wd+5fk0uP+scMfXfPQpMBALy9dvHR5NI3V+aTVz6wToQlyRPzF+eTVz6Qv772ZoHJAADWr12E2A0PP5+//HXpeo+/uHhZrr9/7macCABg49pFiE1/5IVm2QMAsDm1ixB79fUVG93zytLlm2ESAIDGaxchtu3W3RuxZ6vNMAkAQOO1ixA7/H0DN7rnE2M2vgcAYHNqFyH2sZH9ssewbdZ7fJcBvXPYe4UYANC6tIsQ69SxQ66a8P4c/cG6bNWl49r1Lp065ND3bptpJ+yW7m9ZBwBoDSrVarW6uU+6ePHi9O7dO4sWLUpNTU3zvvayFXnk2YWpr1Yzets+6dujS7O+PgBsqVry9/eWqt3cWX+Nmm6dM3aHd5ceAwBgo9rFR5MAAG2REAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFCIEAMAKESIAQAUIsQAAAoRYgAAhQgxAIBChBgAQCFCDACgECEGAFBIp9IDNJflK+vzhxcWZVV9NcP79Urv7p1LjwQAsEFtPsRW1Vdz6Z2z84Pfzs3fli5PknTv3DEH/WNtzth3p9R0E2QAQOvU5kPsCz+emRmPzmuw9saKVfnRg8/lsecX5Sef+WB6dG3zPyYA0A616WvEfvPUX9eJsLeaNW9xrr9/7macCACg8dp0iP3XQ881yx4AgBLadIg9+7elG90ztxF7AABKaNMhVtOIv4xszB4AgBLadIjtP7p2o3sOaMQeAIAS2nSIHTC6NkPf3WO9x3t165Tj/mnwZpwIAKDx2nSIdevcMdcfv1tGD+yzzrEBfbpn6rFjUrfN+kMNAKCkNn+Drf69u+emSXvk93Nfya//9HJW1lez68A++T87/UM6dqiUHg8AYL3afIit8d66vnlvXd/SYwAANFqb/mgSAKAtE2IAAIUIMQCAQoQYAEAhQgwAoBAhBgBQiBADAChEiAEAFCLEAAAKKXJn/Wq1miRZvHhxidMDAO/Amt/ba36Ps+mKhNiSJUuSJAMHDixxegBgEyxZsiS9e/cuPUa7UKkWyNr6+vrMmzcvvXr1SqXii7kBoC2oVqtZsmRJamtr06GDq5uaQ5EQAwDAxfoAAMUIMQCAQoQYAEAhQgwAoBAhBmzQxIkTU6lUUqlU0qVLlwwbNizf+MY3snLlyiSr/4pqypQp2W233dKzZ8/06dMn73vf+3LRRRfl9ddfT5LMmjUrhxxySAYNGpRKpZKLLrqo4E8E0HoIMWCjxo8fn/nz52f27Nk59dRTc9ZZZ+XCCy9Mknz605/O5MmTc+CBB+auu+7KzJkz82//9m+56aabcscddyRJXn/99QwZMiTnn39++vXrV/JHAWhV3L4C2KCJEydm4cKFmT59+tq1ffbZJ0uWLMkXvvCFHHHEEZk+fXoOPPDABs+rVqtZvHjxOjd9HDRoUCZPnpzJkydvhukBWjfviAFN1r179yxfvjzTpk3L8OHD14mwJKlUKu68DbARQgxotGq1ml/+8pe5/fbbs+eee2b27NkZPnx46bEA2iwhBmzULbfckp49e6Zbt2752Mc+liOOOCJnnXWWL/4F2ERFvvQbaFvGjRuXK664Il26dEltbW06dVr9n44ddtghTz75ZOHpANou74gBG9WjR48MGzYs22233doIS5Kjjjoqf/7zn3PTTTet85xqtZpFixZtzjEB2hwhBrxjhx9+eI444ogceeSROe+88/K73/0uc+fOzS233JK99tord911V5Jk+fLlmTlzZmbOnJnly5fnhRdeyMyZM/PUU08V/gkAynL7CmCD3u72FW9VX1+fKVOm5Oqrr86sWbPSqVOnbL/99jn66KNzwgknpHv37nnmmWcyePDgdZ47duzY3H333S37AwC0YkIMAKAQH00CABQixAAAChFiAACFCDEAgEKEGABAIUIMAKAQIQYAUIgQAwAoRIgBABQixAAAChFiAACFCDEAgEL+P6bgCXfkkvNWAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 7.61 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Condition RS vs C\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "geneID                                                                      \n",
      "ENSG00000000003  122.089924        0.079993  0.235417  0.339792  0.734013   \n",
      "ENSG00000000419  186.335951       -0.141822  0.216820 -0.654101  0.513047   \n",
      "ENSG00000000457   22.572713        0.225161  0.481875  0.467261  0.640313   \n",
      "ENSG00000000460   62.835468       -0.276412  0.279343 -0.989508  0.322415   \n",
      "ENSG00000000938    0.304707        1.443372  4.468958  0.322977  0.746712   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSG00000310523    0.903446        2.911071  3.339188  0.871790  0.383323   \n",
      "ENSG00000310527    0.334816        1.546671  4.662615  0.331718  0.740103   \n",
      "ENSG00000310532    0.167408        0.771330  4.803356  0.160581  0.872423   \n",
      "ENSG00000310537    0.172429        0.771330  4.803356  0.160581  0.872423   \n",
      "ENSG00000310539    0.132278        0.771330  4.803356  0.160581  0.872423   \n",
      "\n",
      "                     padj  \n",
      "geneID                     \n",
      "ENSG00000000003  0.999156  \n",
      "ENSG00000000419  0.999156  \n",
      "ENSG00000000457  0.999156  \n",
      "ENSG00000000460  0.999156  \n",
      "ENSG00000000938  0.999156  \n",
      "...                   ...  \n",
      "ENSG00000310523  0.999156  \n",
      "ENSG00000310527  0.999156  \n",
      "ENSG00000310532  0.999156  \n",
      "ENSG00000310537  0.999156  \n",
      "ENSG00000310539  0.999156  \n",
      "\n",
      "[25045 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "# Run combined test\n",
    "combined_results = run_combined_test(count_matrix, metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "321a4375-d68b-4eeb-9660-f57271415c7f",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "LhTPw_YTX2Ob",
    "outputId": "7f345894-0f33-40c9-9db2-39b15b0691ce"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "geneID                                                                      \n",
      "ENSG00000000003  122.089924        0.079993  0.235417  0.339792  0.734013   \n",
      "ENSG00000000419  186.335951       -0.141822  0.216820 -0.654101  0.513047   \n",
      "ENSG00000000457   22.572713        0.225161  0.481875  0.467261  0.640313   \n",
      "ENSG00000000460   62.835468       -0.276412  0.279343 -0.989508  0.322415   \n",
      "ENSG00000001036  223.518204       -0.006949  0.180606 -0.038478  0.969307   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSG00000310385   36.231619       -0.014821  0.371207 -0.039927  0.968151   \n",
      "ENSG00000310431   10.435667       -0.990067  0.729537 -1.357117  0.174744   \n",
      "ENSG00000310441   30.178809        0.336015  0.426458  0.787921  0.430743   \n",
      "ENSG00000310517  231.406618        0.175411  0.167736  1.045755  0.295674   \n",
      "ENSG00000310521   16.187346        0.582236  0.586035  0.993517  0.320458   \n",
      "\n",
      "                     padj    Symbol  \n",
      "geneID                               \n",
      "ENSG00000000003  0.999156    TSPAN6  \n",
      "ENSG00000000419  0.999156      DPM1  \n",
      "ENSG00000000457  0.999156     SCYL3  \n",
      "ENSG00000000460  0.999156  C1orf112  \n",
      "ENSG00000001036  0.999156     FUCA2  \n",
      "...                   ...       ...  \n",
      "ENSG00000310385  0.999156       NaN  \n",
      "ENSG00000310431  0.999156       NaN  \n",
      "ENSG00000310441  0.999156       NaN  \n",
      "ENSG00000310517  0.999156       NaN  \n",
      "ENSG00000310521  0.999156       NaN  \n",
      "\n",
      "[10806 rows x 7 columns]\n",
      "Empty DataFrame\n",
      "Columns: [baseMean, log2FoldChange, lfcSE, stat, pvalue, padj, Symbol]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "# filter out values\n",
    "combined_results = combined_results[combined_results.baseMean >= 10]\n",
    "\n",
    "# Access combined test results\n",
    "print(combined_results)\n",
    "\n",
    "# Filter significant genes\n",
    "sigs = combined_results[(combined_results['padj'] < 0.5) & (abs(combined_results['log2FoldChange']) > 2.0)]\n",
    "\n",
    "print(sigs)\n",
    "\n",
    "# Save results\n",
    "#combined_results.to_csv(\"RS_vs_C_significant_genes.csv\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26934be3-92e5-4d2b-af1a-113bc6ee6661",
   "metadata": {
    "id": "qgq56N_HX3Mn"
   },
   "source": [
    "## Run each treatment against all the controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "fd0c0b5f-c691-4f41-931f-c1efc0c82cd7",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "0ugHueNNfmAl",
    "outputId": "19612bd0-c6ad-4c9b-a10c-f87750328d63"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 17.58 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.79 seconds.\n",
      "\n",
      "/usr/local/lib/python3.10/dist-packages/pydeseq2/dds.py:497: UserWarning: As the residual degrees of freedom is less than 3, the distribution of log dispersions is especially asymmetric and likely to be poorly estimated by the MAD.\n",
      "  self.fit_dispersion_prior()\n",
      "Fitting MAP dispersions...\n",
      "... done in 28.41 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 23.26 seconds.\n",
      "\n",
      "Calculating cook's distance...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAmMAAAGvCAYAAAAXCtIvAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAmFUlEQVR4nO3deXRU9R338c9NJhtJJmFPIpAQAkFCQVFxqwUkUFbRxyouCLRqq6ZCCoqAp61YF1qtAq2lYlnL2j7KqgKiIOWhAgpRAcGAASlkQSGTkBhCMvf5g5IyJmSBmfwy4f06x9OTe38zv2/sOeZ97tzcWLZt2wIAAIARAaYHAAAAuJwRYwAAAAYRYwAAAAYRYwAAAAYRYwAAAAYRYwAAAAYRYwAAAAYRYwAAAAYRYwAAAAYRYwC8at68ebIsS4cOHao41rt3b/Xu3btWrx89erQSEhJ8MhsANETEGNAIHDx4UL/4xS+UmJio0NBQOZ1O3XzzzZo+fbq+++470+NVcuzYMT3zzDPKyMgwPQoAGOcwPQCAS/P222/rrrvuUkhIiEaOHKmuXbuqtLRUW7Zs0ZNPPqk9e/Zo1qxZRmdcv369x9fHjh3TlClTlJCQoKuuusrj3BtvvCG3212P0wGAWcQY4MeysrJ0zz33KD4+Xh988IFiY2MrzqWlpenAgQN6++23DU54VnBwcK3XBgUF+XASAGh4+JgS8GN/+MMfdOrUKc2ePdsjxM5JSkrS2LFjJUllZWX63e9+pw4dOigkJEQJCQmaPHmyTp8+7fGahIQEDRkyRFu2bFHPnj0VGhqqxMRELViwoNL779mzR7feeqvCwsLUpk0bPffcc1Ve1Tr/nrFNmzbpuuuukyT99Kc/lWVZsixL8+bNk1T1PWNFRUUaP3682rZtq5CQECUnJ+vll1+Wbdse6yzL0i9/+UutWLFCXbt2VUhIiFJSUrR27dpa/fsEABMs+/v/NQPgN9q0aaOQkBAdPHiwxrWjR4/W/Pnz9ZOf/ER9+vTRtm3btGDBAt1+++1avnx5xbqEhASFhoYqPz9fDz74oOLi4jRnzhzt2rVLn3/+uVJSUiRJOTk56tatm8rKyjR27FiFh4dr1qxZCgsL02effaasrKyKqDo/xHJzczVr1iz95je/0c9//nPdcsstkqSbbrpJiYmJGj16tDZt2lTxCwC2bSs1NVUbN27Ugw8+qKuuukrr1q3T6tWrlZ6erldffbVidsuy1L17d+Xl5emxxx5TZGSkZsyYoZycHH399ddq3ry5F/6tA4CX2QD8ksvlsiXZw4YNq3FtRkaGLcl+6KGHPI4/8cQTtiT7gw8+qDgWHx9vS7I3b95ccSwvL88OCQmxx48fX3EsPT3dlmRv27bNY11UVJQtyc7Kyqo43qtXL7tXr14VX+/YscOWZM+dO7fSrKNGjbLj4+Mrvl6xYoUtyX7uuec81v3kJz+xLcuyDxw4UHFMkh0cHOxx7NNPP7Ul2X/6058u/C8IAAziY0rATxUUFEiSIiMja1z7zjvvSJLGjRvncXz8+PGSVOm+si5dulRcsZKkli1bKjk5WV999ZXHe95www3q2bOnx7r777+/jt9JzbMHBgZqzJgxlWa3bVvvvvuux/HU1FR16NCh4utu3brJ6XR6zA4ADQkxBvgpp9MpSSosLKxx7eHDhxUQEKCkpCSP4zExMYqOjtbhw4c9jrdr167SezRt2lQnT570eM+OHTtWWpecnFyr+Wvr8OHDiouLqxSdV155ZcX589VmdgBoSIgxwE85nU7FxcVp9+7dtX6NZVm1WhcYGFjlcdsPbjH159kBXJ6IMcCPDRkyRAcPHtS///3vatfFx8fL7XYrMzPT43hubq7y8/MVHx9f573j4+MrvZ8k7d+/v8bX1jYKz+1z7NixSlcA9+3bV3EeAPwZMQb4sQkTJig8PFwPPfSQcnNzK50/ePCgpk+frkGDBkmSpk2b5nH+lVdekSQNHjy4znsPGjRIH330kbZv315x7Pjx41q0aFGNrw0PD5ck5efn12qf8vJy/fnPf/Y4/uqrr8qyLA0cOLBugwNAA8NDXwE/1qFDBy1evFjDhw/XlVde6fEE/q1bt+qf//ynRo8erbFjx2rUqFGaNWuW8vPz1atXL23fvl3z58/X7bffrj59+tR57wkTJujvf/+7BgwY4PFoi/j4eH322Wc1zh0dHa2//vWvioyMVHh4uK6//nq1b9++0tqhQ4eqT58+evrpp3Xo0CF1795d69ev18qVK5Wenu5xsz4A+CNiDPBzt912mz777DO99NJLWrlypWbOnKmQkBB169ZNf/zjH/Xwww9Lkv72t78pMTFR8+bN0/LlyxUTE6NJkybpt7/97UXtGxsbq40bN+rxxx/X1KlT1bx5cz3yyCOKi4vTgw8+WO1rg4KCNH/+fE2aNEmPPPKIysrKNHfu3CpjLCAgQKtWrdJvfvMbLVu2THPnzlVCQoJeeumlit8GBQB/xkNfAQAADOKeMQAAAIOIMQAAAIOIMQAAAIOIMQAAAIOIMQAAAIOIMQAAAIOMPGfM7Xbr2LFjioyMrNOfRQEAAObYtq3CwkLFxcUpIIDrOd5iJMaOHTumtm3bmtgaAABcoiNHjqhNmzamx2g0jMRYZGSkpLP/ZzqdThMjAACAOiooKFDbtm0rfo7DO4zE2LmPJp1OJzEGAICf4RYj7+IDXwAAAIOIMQAAAIOIMQAAAIOIMQAAAIOIMQAAAIOIMQAAAIOIMQAAAIOIMQAAAIOMPPT1smLbUvanUukpqVmi5IwzPREAAGhAiDFfylgsbX5ZOnHw7NdWoNTpx1L/56TmHczOBgAAGgQ+pvSV/zddWvHo/0JMkuxyaf870pwfSye+MjcbAABoMIgxXzh1XPrguQufLzouvf+7+psHAAA0WMSYL3y6RCovrX7NF6ul4hP1Mw8AAGiwiDFfqM1HkO4zkus/vp8FAAA0aMSYL4RF125daJRPxwAAAA0fMeYLXe+seU2bnlLTeN/PAgAAGjRizBdifiB1HnLh81aA1Oup+psHAAA0WMSYr/yfN6SUOyRZnsdDoqQ7ZkkdU42MBQAAGhYe+uorwU2ku+ZJfQ5Ie1dIpwulFp3OBlpwE9PTAQCABoIY87UWSdKPnjA9BQAAaKD4mBIAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgYgwAAMAgh+kBGqvdR106UVSq2KhQdWwdaXocAADQQBFjXvbe3ly9vG6/9ucWVhy7qm20Jg+6Uj3bNzM4GQAAaIj4mNKLVn96TL/4+8ceISZJGUfyNWL2Nv374LeGJgMAAA0VMeYlZ8rdmrJ6r9x21edLy9x6ds3e+h0KAAA0eMSYl7z/RZ6+OXW62jVfZBfo0yP59TMQAADwC8SYl/znZHGt1h2p5ToAAHB5IMa8JCosqFbrosOCfTwJAADwJ8SYl/RPiVFYUGC1a2Kcobohkd+oBAAA/0OMeUlUWJAeuqV9tWvG9O0oRyD/ygEAwP/wnDEvGtevk86U25qzJUul5e6K402CAzWuXyfdd307g9MBAICGyLJt+wIPY/CdgoICRUVFyeVyyel01vf2PvftqdNa81m2vj11WrHRYRrSLVaRobW7pwwAgIaqsf/8NoUrYz7QPCJEo25KMD0GAADwA9zABAAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYBAxBgAAYJDD9AD+4kx2tgrf2yB3cZGC2ycq8tY+soKCTI8FAAD8HDFWA3dpqXKmTJFrxUqpvLziuKNlS8U8O0WRffoYnA4AAPg7PqasQfbEiXK9+ZZHiElS2fHj+s+YsSrescPQZAAAoDEgxqpR8uWXKnjn3QsvOHNGx//8Wv0NBAAAvG7z5s0aOnSo4uLiZFmWVqxYUeW6JUuWKDAwUGlpaV7dnxirRsHqNTWuKd62TWdyc+thGgAAGr9yt61/H/xWKzOO6t8Hv1W52/b5nkVFRerevbtee636CyyzZ8/WhAkTtGTJEpWUlHhtf+4Zq0Z5/slarnMpqHVrH08DAEDjtnZ3tqas3qts1/9CJzYqVL8d2kUDusb6bN+BAwdq4MCB1a7JysrS1q1b9eabb2rjxo166623dN9993llf66MVSMoLq7mRQ6HHK1a+n4YAAAasbW7s/Xowp0eISZJOa4SPbpwp9buzjY02Vlz587V4MGDFRUVpREjRmj27Nlee29irBpRd9whBQZWuyYyNVWOpk3raSIAABqfcretKav3qqoPJM8dm7J6b718ZFkVt9utefPmacSIEZKke+65R1u2bFFWVpZX3p8Yq0ZQTIya//zhC54PjIpSyzFj6nEiAAAan+1ZJypdETufLSnbVaLtWSfqb6jzvPfeeyoqKtKgQYMkSS1atFC/fv00Z84cr7w/MVaDVmPHqvXkyXK0auVxvMmNNyh+0UKFJLY3NBkAAI1DXmHtboav7Tpvmz17tk6cOKGwsDA5HA45HA698847mj9/vtxu9yW/Pzfw10KzkQ+o6X33qnjnTrmLihTSvr2CExJMjwUAQKPQKjLUq+u86dtvv9XKlSu1dOlSpaSkVBwvLy/XD3/4Q61fv14DBgy4pD2IsVqyHA6F9+xpegwAABqdnu2bKTYqVDmukirvG7MkxUSFqmf7Zj7Z/9SpUzpw4EDF11lZWcrIyFCzZs301ltvqXnz5rr77rtlWZbH6wYNGqTZs2cTYwAAwL8FBlj67dAuenThTlmSR5Cdy5/fDu2iwACrildfuo8//lh9zvvzhuPGjZMkjRo1Sjt37tQdd9xRKcQk6c4779QDDzygb775Ri1atLjo/S3btuv9VxMKCgoUFRUll8slp9NZ39sDAICL4Ouf36aeM2YaV8YAAECDMKBrrPp1idH2rBPKKyxRq8izH0366opYQ0GMAQCABiMwwNKNHZqbHqNe8WgLAAAAg4gxAAAAg4gxAAAAg4gxAAAAg4gxAAAAg4gxAAAAg4gxAAAAg4gxAAAAg4gxAABw2XrxxRd13XXXKTIyUq1atdLtt9+u/fv3e6xJSEiQZVmyLEvh4eHq0aOH/vnPf3ptBmIMAAA0HO5yKetf0uf/9+z/ust9ut2HH36otLQ0ffTRR3rvvfd05swZ9e/fX0VFRR7rnn32WWVnZ2vXrl267rrrNHz4cG3dutUrM/DnkAAAQMOwd5W09imp4Nj/jjnjpAG/l7rc5pMt165d6/H1vHnz1KpVK33yySf60Y9+VHE8MjJSMTExiomJ0WuvvaaFCxdq9erVuummmy55Bq6MAQAA8/aukv4x0jPEJKkg++zxvavqZQyXyyVJatas2QXXOBwOBQUFqbS01Ct7EmMAAMAsd/nZK2Kyqzj532NrJ/r8I0u326309HTdfPPN6tq1a5VrSktL9eKLL8rlcunWW2/1yr7EGAAAMOvw1spXxDzYUsHRs+t8KC0tTbt379bSpUsrnXvqqacUERGhJk2a6Pe//72mTp2qwYMHe2Vf7hkDAABmncr17rqL8Mtf/lJr1qzR5s2b1aZNm0rnn3zySY0ePVoRERFq3bq1LMvy2t7EGAAAMCuitXfX1YFt23r88ce1fPlybdq0Se3bt69yXYsWLZSUlOT1/SViDAAAmBZ/09nfmizIVtX3jVlnz8df+m8ufl9aWpoWL16slStXKjIyUjk5OZKkqKgohYWFeX2/qnDPGAAAMCsg8OzjKyRJ3//4779fD5h6dp2XzZw5Uy6XS71791ZsbGzFP8uWLfP6XhfClTEAAGBel9ukuxdc4DljU332nDHbrupKnKdDhw75ZO9ziDEAANAwdLlN6jz47G9Nnso9e49Y/E0+uSLWkBBjAACg4QgIlNrfYnqKesU9YwAAAAYRYwAAAAYRYwAAAAYRYwAAAAYRYwAAAAbVOcays7O1cOFCvfPOOyotLfU4V1RUpGeffdZrwwEAADR2ll2bp539144dO9S/f3+53W6dOXNGV1xxhVasWKGUlBRJUm5uruLi4lReXl7t+xQUFCgqKkoul0tOp/PSvgMAAFAv+PntG3W6MjZ58mTdcccdOnnypHJzc9WvXz/16tVLu3bt8tV8AAAAjVqdYuyTTz7RxIkTFRAQoMjISP3lL3/RE088ob59+2rHjh2+mhEAAMAnZs6cqW7dusnpdMrpdOrGG2/Uu+++W3E+ISFB06ZN8+kMdX4Cf0lJicfXEydOlMPhUP/+/TVnzhyvDQYAAC4/5e5y7czbqePFx9WySUv1aNVDgT78c0ht2rTR1KlT1bFjR9m2rfnz52vYsGHatWtXxW1YvlanGOvatau2bt2qbt26eRx/4okn5Ha7de+993p1OAAAcPnYcHiDpm6fqtzi3IpjrZu01sSeE5Uan+qTPYcOHerx9fPPP6+ZM2fqo48+qrcYq9PHlCNHjtSWLVuqPDdhwgRNmTJF7dq188pgAADg8rHh8AaN2zTOI8QkKa84T+M2jdOGwxt8PkN5ebmWLl2qoqIi3XjjjT7f75w6xdhDDz2khQsXXvD8U089paysrEseCgAAXD7K3eWaun2qbFV+wMO5Y7/f/nuVu6t/WsPF+vzzzxUREaGQkBA98sgjWr58ubp06eKTvapSpxgrKSnRqlWrVFhYWOlcQUGBVq1apdOnT3ttOAAA0PjtzNtZ6YrY+WzZyinO0c68nT7ZPzk5WRkZGdq2bZseffRRjRo1Snv37vXJXlWpU4y9/vrrmj59uiIjIyudczqdmjFjht544w2vDQcAABq/48XHvbquroKDg5WUlKRrrrlGL774orp3767p06f7ZK+q1CnGFi1apPT09AueT09P14IFCy51JgAAcBlp2aSlV9ddKrfbXa+f9NXptykzMzPVvXv3C57v1q2bMjMzL3koAABw+ejRqodaN2mtvOK8Ku8bs2SpdZPW6tGqh9f3njRpkgYOHKh27dqpsLBQixcv1qZNm7Ru3bqKNUePHlVGRobH6+Lj49W0aVOvzFCnK2NlZWU6fvzClwiPHz+usrKySx4KAABcPgIDAjWx50RJZ8PrfOe+fqrnUz553lheXp5Gjhyp5OTkiofYr1u3Tv369atY8/LLL+vqq6/2+Oftt9/22gx1ujKWkpKiDRs26Jprrqny/Pr16+vtmRwAAKDxSI1P1Su9X6nyOWNP9XzKZ88Zmz17drXnDx065JN9z1enGPvZz36mcePGKSUlRUOGDPE4t3r1aj3//PN65ZVXvDogAAC4PKTGp6pP2z71+gT+hsCybbvyh7PVGDFihBYvXqzOnTsrOTlZkrRv3z59+eWXuvvuu7VkyZIa34O/+g4AgP/h57dv1OmeMUlauHChli1bpk6dOunLL7/U/v37lZycrCVLltQqxAAAAPA/dfqYsry8XC+//LJWrVql0tJSDRkyRM8884zCwsJ8NR8AAECjVqcrYy+88IImT56siIgIXXHFFZoxY4bS0tJ8NRsAAECjV6cYW7Bggf7yl79o3bp1WrFihVavXq1FixbJ7Xb7aj4AAIBGrU4x9vXXX2vQoEEVX6empsqyLB07dszrgwEAAFwO6vzQ19DQUI9jQUFBOnPmjFeHAgAAuFzU6QZ+27Y1evRohYSEVBwrKSnRI488ovDw8Ipjb731lvcmBAAAaMTqFGOjRo2qdGzEiBFeGwYAAOByU6cYmzt3rq/mAAAAMG7q1KmaNGmSxo4dq2nTplUc37Vrl1544QVt3rxZLpdLbdu2Ve/evfXkk0+qU6dOl7RnnR/6CgAA4Ct2ebmKtm2Xa83bKtq2XXZ5eb3tvWPHDr3++uvq1q2bx/E1a9bohhtu0OnTp7Vo0SJ98cUXWrhwoaKiovTrX//6kvet05UxAAAAXylYv165L7yospycimOOmBi1njxJzv79fbr3qVOndP/99+uNN97Qc889V3G8uLhYP/3pTzVo0CAtX7684nj79u11/fXXKz8//5L35soYAACNQElZiZZnLtfTW57W01ue1luZb6mkrMT0WLVWsH69jo5N9wgxSSrLzdXRsekqWL/ep/unpaVp8ODBSk1N9Ti+bt06ffPNN5owYUKVr4uOjr7kvbkyBgCAn/v0+Kca88EYnSg5UXFs1cFVmvbJNE2/dbqubnW1welqZpeXK/eFFyXbruKkLVmWcl94UZF9+8oKDPT6/kuXLtXOnTu1Y8eOSucyMzMlSZ07d/b6vudwZQwAAD+WW5SrRzc86hFi55w8fVJpG9KUfSrbwGS1V/zxJ5WuiHmwbZXl5Kj440+8vveRI0c0duxYLVq0qNKzVM9uXUUgehkxBgCAH1u2f5kKSwsveL7wTKGW7l9ajxPVXdnx415dVxeffPKJ8vLy1KNHDzkcDjkcDn344YeaMWOGHA6HkpKSJEn79u3z+t7nEGMAAPix979+v8Y1Gw5vqIdJLp6jZUuvrquLvn376vPPP1dGRkbFP9dee63uv/9+ZWRkaMCAAWrRooX+8Ic/VPl6b9zAzz1jAAD4sdrcpP9d2Xf1MMnFa3LtNXLExKgsN7fq+8YsS47WrdXk2mu8vndkZKS6du3qcSw8PFzNmzevOP63v/1Nd911l2677TaNGTNGSUlJ+uabb/SPf/xDX3/9tZYuvbQrj1wZAwDAj3WI7lDjmqTopHqY5OJZgYFqPXnSf7+wvnfy7NetJ0/yyc37tTFs2DBt3bpVQUFBuu+++9S5c2fde++9crlcHo/BuFiWXR93pn1PQUGBoqKi5HK55HQ663t7AAAajU1HNunxDx6vds2rvV9VanxqtWtqw9c/v00+Z8wkPqYEAMCP9WrTS0MSh2jNV2uqPD8wYaD6tutbz1NdHGf//ors2/fsb1cePy5Hy5Zqcu01xq6I1RdiDAAAP2ZZlp7/4fNKaZ6ixfsW60jhEUlSm4g2uqfzPXqgywOyvv/RXwNmBQYq/PqepseoV3xMCQBAI2HbtnKKcmTLVmx4rNcjjJ/fvsGVMQAAGgnLshQbEWt6DNQRv00JAABgEDEGAABgEDEGAABgEDEGAABgEDEGAABgEDEGAABgEDEGAAAuezk5OXr88ceVmJiokJAQtW3bVkOHDtX777/v8715zhgAAGgw3G5b2Zn5Kio4rXBniGI7RisgwLd/QeDQoUO6+eabFR0drZdeekk/+MEPdObMGa1bt05paWnat2+fT/cnxgAAQINwcFee/rUsU0X5pyuOhUeH6JbhHdXh6lY+2/exxx6TZVnavn27wsPDK46npKToZz/7mc/2PYePKQEAgHEHd+Vp7eu7PUJMkoryT2vt67t1cFeeT/Y9ceKE1q5dq7S0NI8QOyc6Oton+56PGAMAAEa53bb+tSyz2jVb/pEpt9v7f077wIEDsm1bnTt39vp71xYxBgAAjMrOzK90Rez7Tp08rezMfK/vbdveD7y6IsYAAIBRRQXVh1hd19VFx44dZVmWz2/Srw4xBgAAjAp3hnh1XV00a9ZMP/7xj/Xaa6+pqKio0vn8/Hyv7/l9xBgAADAqtmO0wqOrD62Ipmcfc+ELr732msrLy9WzZ0+9+eabyszM1BdffKEZM2boxhtv9Mme5yPGAACAUQEBlm4Z3rHaNT+8u6PPnjeWmJionTt3qk+fPho/fry6du2qfv366f3339fMmTN9suf5LNvAnWsFBQWKioqSy+WS0+ms7+0BAMBF8PXP76qeMxbRNEQ/vNu3zxkzjYe+AgCABqHD1a3UvnvLen8Cv2nEGAAAaDACAixdkdzU9Bj1invGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADCLGAAAADHKYHsDb8g4X6D/7Tsq2bcUmRSsuKdr0SAAAABfUaGLs1MkSrXtjj3K+cnkcb9E2Qv0fTFHTmHBDkwEAAFxYo/iYsrSkTCunZVQKMUn65sgprZyWoeKCUgOTAQAAVK9RxNj+j3KUn1t8wfNF+ae1e/PRepwIAACgdhpFjH25PafmNdtqXgMAAFDfGkWMfXfqTI1rSopqXgMAAFDfGkWMOZuH1rgmshZrAAAA6lujiLErb46rcU2XWqwBAACob40ixjpc3VJtOje94PmW7SLV+abYepwIAACgdhpFjAUEBmjwY930g15XyBESWHE80BGgzjfGaNivrlZQcGA17wAAAGCGZdu2Xd+bFhQUKCoqSi6XS06n06vvffq7MuV+5ZJtS60SIhUWEezV9wcA4HLly5/fl7NG8wT+c0LCHGqX0tz0GAAAALXSKD6mBAAA8FfEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEHEGAAAgEEO0wN4S3mZW8e/LpS73FbzK8IV0iTI9EgAAAA18vsYc7ttffzOIe3+8D/6rvCMJMkRHKBOPWN0051JCgnz+28RAAA0Yn5fKhvm7lXmjlyPY2Wlbu3dckx5hwt0x/geCg71+28TAAA0Un59z9iRfScqhdj5vjlySrs3H63HiQAAAOrGr2Psiy3HalyztxZrAAAATPHrGHMd/67GNQW1WAMAAGCKX8dYSJOa7wULrsUaAAAAU/w6xjpe17rGNZ2urXkNAACAKf4dY9e2VtOYJhc8HxzmUPfUtvU4EQAAQN34dYw5ggN129ir1CrBWelcZLNQDR3TXVEtLxxrAAAApvn9DVURTUN118RrlX3Qpa/3fCu321brBKcSurVQQIBlejwAAIBq+X2MnRPbIUqxHaJMjwEAAFAnfv0xJQAAgL8jxgAAAAwixgAAAAwixgAAAAwixgAAAAwixgAAAAwixgAAAAwixgAAAAwixgAAAAwy8gR+27YlSQUFBSa2BwAAF+Hcz+1zP8fhHUZirLCwUJLUtm1bE9sDAIBLUFhYqKgo/gSht1i2gbx1u906duyYIiMjZVn8MW8AAPyBbdsqLCxUXFycAgK408lbjMQYAAAAziJrAQAADCLGAAAADCLGAAAADCLGAAAADCLGAFRr9OjRsixLlmUpODhYSUlJevbZZ1VWVibp7G9XzZo1S9dff70iIiIUHR2ta6+9VtOmTVNxcbEkac+ePbrzzjuVkJAgy7I0bdo0g98RADQsxBiAGg0YMEDZ2dnKzMzU+PHj9cwzz+ill16SJD3wwANKT0/XsGHDtHHjRmVkZOjXv/61Vq5cqfXr10uSiouLlZiYqKlTpyomJsbktwIADQ6PtgBQrdGjRys/P18rVqyoONa/f38VFhbqV7/6lYYPH64VK1Zo2LBhHq+zbVsFBQWVHgyZkJCg9PR0paen18P0ANDwcWUMQJ2FhYWptLRUixYtUnJycqUQkyTLsnhCNwDUAjEGoNZs29aGDRu0bt063XrrrcrMzFRycrLpsQDArxFjAGq0Zs0aRUREKDQ0VAMHDtTw4cP1zDPP8MeCAcALjPyhcAD+pU+fPpo5c6aCg4MVFxcnh+Psfzo6deqkffv2GZ4OAPwbV8YA1Cg8PFxJSUlq165dRYhJ0n333acvv/xSK1eurPQa27blcrnqc0wA8EvEGICLdvfdd2v48OG699579cILL+jjjz/W4cOHtWbNGqWmpmrjxo2SpNLSUmVkZCgjI0OlpaU6evSoMjIydODAAcPfAQCYx6MtAFSrqkdbnM/tdmvWrFmaM2eO9uzZI4fDoY4dO2rkyJF6+OGHFRYWpkOHDql9+/aVXturVy9t2rTJt98AADRwxBgAAIBBfEwJAABgEDEGAABgEDEGAABgEDEGAABgEDEGAABgEDEGAABgEDEGAABgEDEGAABgEDEGAABgEDEGAABgEDEGAABgEDEGAABg0P8HwQH2IqufWoYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 6.13 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Condition 1A vs C\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "geneID                                                                      \n",
      "ENSG00000000003  122.089924        0.422939  0.403215  1.048916  0.294217   \n",
      "ENSG00000000419  186.335951        0.119526  0.339289  0.352285  0.724625   \n",
      "ENSG00000000457   22.572713        0.109584  0.886316  0.123640  0.901601   \n",
      "ENSG00000000460   62.835468       -0.430385  0.548798 -0.784231  0.432905   \n",
      "ENSG00000000938    0.304707        2.544266  6.247868  0.407221  0.683845   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSG00000310523    0.903446        0.101758  6.280527  0.016202  0.987073   \n",
      "ENSG00000310527    0.334816        0.101736  6.947422  0.014644  0.988316   \n",
      "ENSG00000310532    0.167408        0.101724  6.947428  0.014642  0.988318   \n",
      "ENSG00000310537    0.172429        2.544239  6.247871  0.407217  0.683849   \n",
      "ENSG00000310539    0.132278        0.101724  6.947428  0.014642  0.988318   \n",
      "\n",
      "                     padj  \n",
      "geneID                     \n",
      "ENSG00000000003  0.999973  \n",
      "ENSG00000000419  0.999973  \n",
      "ENSG00000000457  0.999973  \n",
      "ENSG00000000460  0.999973  \n",
      "ENSG00000000938  0.999973  \n",
      "...                   ...  \n",
      "ENSG00000310523  0.999973  \n",
      "ENSG00000310527  0.999973  \n",
      "ENSG00000310532  0.999973  \n",
      "ENSG00000310537  0.999973  \n",
      "ENSG00000310539  0.999973  \n",
      "\n",
      "[25045 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 8.45 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Condition 2P vs C\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "geneID                                                                      \n",
      "ENSG00000000003  122.089924        0.074891  0.407199  0.183919  0.854077   \n",
      "ENSG00000000419  186.335951       -0.068959  0.340854 -0.202312  0.839673   \n",
      "ENSG00000000457   22.572713        0.318489  0.877325  0.363023  0.716588   \n",
      "ENSG00000000460   62.835468       -0.390557  0.546729 -0.714351  0.475010   \n",
      "ENSG00000000938    0.304707        0.059051  6.947795  0.008499  0.993219   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSG00000310523    0.903446        4.086663  5.369149  0.761138  0.446575   \n",
      "ENSG00000310527    0.334816        3.501695  6.164244  0.568066  0.569990   \n",
      "ENSG00000310532    0.167408        2.501704  6.248084  0.400395  0.688865   \n",
      "ENSG00000310537    0.172429        0.059051  6.947795  0.008499  0.993219   \n",
      "ENSG00000310539    0.132278        0.059036  6.947816  0.008497  0.993220   \n",
      "\n",
      "                    padj  \n",
      "geneID                    \n",
      "ENSG00000000003  0.99994  \n",
      "ENSG00000000419  0.99994  \n",
      "ENSG00000000457  0.99994  \n",
      "ENSG00000000460  0.99994  \n",
      "ENSG00000000938  0.99994  \n",
      "...                  ...  \n",
      "ENSG00000310523  0.99994  \n",
      "ENSG00000310527  0.99994  \n",
      "ENSG00000310532  0.99994  \n",
      "ENSG00000310537  0.99994  \n",
      "ENSG00000310539  0.99994  \n",
      "\n",
      "[25045 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 5.77 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Condition 3L vs C\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "geneID                                                                      \n",
      "ENSG00000000003  122.089924       -0.361927  0.408642 -0.885683  0.375788   \n",
      "ENSG00000000419  186.335951       -0.753888  0.345216 -2.183818  0.028976   \n",
      "ENSG00000000457   22.572713       -0.343233  0.889108 -0.386042  0.699466   \n",
      "ENSG00000000460   62.835468       -0.181457  0.535458 -0.338883  0.734698   \n",
      "ENSG00000000938    0.304707        2.161926  6.248076  0.346015  0.729332   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSG00000310523    0.903446       -0.280755  6.280844 -0.044700  0.964346   \n",
      "ENSG00000310527    0.334816       -0.280754  6.947814 -0.040409  0.967767   \n",
      "ENSG00000310532    0.167408       -0.280754  6.947814 -0.040409  0.967767   \n",
      "ENSG00000310537    0.172429       -0.280739  6.947793 -0.040407  0.967769   \n",
      "ENSG00000310539    0.132278        2.161914  6.248083  0.346012  0.729333   \n",
      "\n",
      "                     padj  \n",
      "geneID                     \n",
      "ENSG00000000003  0.999357  \n",
      "ENSG00000000419  0.999357  \n",
      "ENSG00000000457  0.999357  \n",
      "ENSG00000000460  0.999357  \n",
      "ENSG00000000938  0.999357  \n",
      "...                   ...  \n",
      "ENSG00000310523  0.999357  \n",
      "ENSG00000310527  0.999357  \n",
      "ENSG00000310532  0.999357  \n",
      "ENSG00000310537  0.999357  \n",
      "ENSG00000310539  0.999357  \n",
      "\n",
      "[25045 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 8.61 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Condition 4C vs C\n",
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "geneID                                                                      \n",
      "ENSG00000000003  122.089924        0.107656  0.410745  0.262098  0.793246   \n",
      "ENSG00000000419  186.335951        0.012525  0.343000  0.036516  0.970871   \n",
      "ENSG00000000457   22.572713        0.690554  0.873803  0.790285  0.429361   \n",
      "ENSG00000000460   62.835468       -0.129523  0.546729 -0.236906  0.812730   \n",
      "ENSG00000000938    0.304707        0.320081  6.947797  0.046069  0.963255   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSG00000310523    0.903446        3.762736  5.401356  0.696628  0.486036   \n",
      "ENSG00000310527    0.334816        0.320066  6.947818  0.046067  0.963257   \n",
      "ENSG00000310532    0.167408        0.320066  6.947818  0.046067  0.963257   \n",
      "ENSG00000310537    0.172429        0.320081  6.947797  0.046069  0.963255   \n",
      "ENSG00000310539    0.132278        0.320066  6.947818  0.046067  0.963257   \n",
      "\n",
      "                     padj  \n",
      "geneID                     \n",
      "ENSG00000000003  0.999728  \n",
      "ENSG00000000419  0.999728  \n",
      "ENSG00000000457  0.999728  \n",
      "ENSG00000000460  0.999728  \n",
      "ENSG00000000938  0.999728  \n",
      "...                   ...  \n",
      "ENSG00000310523  0.999728  \n",
      "ENSG00000310527  0.999728  \n",
      "ENSG00000310532  0.999728  \n",
      "ENSG00000310537  0.999728  \n",
      "ENSG00000310539  0.999728  \n",
      "\n",
      "[25045 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "# Define treatments\n",
    "treatments = ['1A', '2P', '3L', '4C']\n",
    "\n",
    "# Run tests\n",
    "individual_results = run_individual_tests(count_matrix, treatments, indTreatMetadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "ff3801db-fd4f-4a1d-a915-d48fb031659f",
   "metadata": {
    "id": "vd2JeRIOfp1L"
   },
   "outputs": [],
   "source": [
    "# Adjust as needed\n",
    "padj = 0.5\n",
    "logfold2 = 2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "d546e9a7-1077-4661-8574-0d4b18f1b7da",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "XTDW5ClxfrQf",
    "outputId": "480cf2b8-b532-45cf-bd35-c6e35a803bbf"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "geneID                                                                      \n",
      "ENSG00000000003  122.089924        0.422939  0.403215  1.048916  0.294217   \n",
      "ENSG00000000419  186.335951        0.119526  0.339289  0.352285  0.724625   \n",
      "ENSG00000000457   22.572713        0.109584  0.886316  0.123640  0.901601   \n",
      "ENSG00000000460   62.835468       -0.430385  0.548798 -0.784231  0.432905   \n",
      "ENSG00000001036  223.518204       -0.078081  0.321366 -0.242968  0.808030   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSG00000310385   36.231619        0.321907  0.694752  0.463341  0.643120   \n",
      "ENSG00000310431   10.435667       -0.576514  1.319438 -0.436939  0.662156   \n",
      "ENSG00000310441   30.178809        0.112838  0.783123  0.144087  0.885432   \n",
      "ENSG00000310517  231.406618        0.319912  0.316512  1.010743  0.312140   \n",
      "ENSG00000310521   16.187346        1.209728  1.020748  1.185138  0.235963   \n",
      "\n",
      "                     padj    Symbol  \n",
      "geneID                               \n",
      "ENSG00000000003  0.999973    TSPAN6  \n",
      "ENSG00000000419  0.999973      DPM1  \n",
      "ENSG00000000457  0.999973     SCYL3  \n",
      "ENSG00000000460  0.999973  C1orf112  \n",
      "ENSG00000001036  0.999973     FUCA2  \n",
      "...                   ...       ...  \n",
      "ENSG00000310385  0.999973       NaN  \n",
      "ENSG00000310431  0.999973       NaN  \n",
      "ENSG00000310441  0.999973       NaN  \n",
      "ENSG00000310517  0.999973       NaN  \n",
      "ENSG00000310521  0.999973       NaN  \n",
      "\n",
      "[10806 rows x 7 columns]\n",
      "Empty DataFrame\n",
      "Columns: [baseMean, log2FoldChange, lfcSE, stat, pvalue, padj, Symbol]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "# 1A\n",
    "oneA = individual_results['1A']\n",
    "oneA = oneA[oneA.baseMean >= 10]\n",
    "print(oneA)\n",
    "\n",
    "# Filter significant genes\n",
    "asigs = oneA[(oneA['padj'] < padj) & (abs(oneA['log2FoldChange']) > logfold2)]\n",
    "print(asigs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "80c78734-3619-4168-9cd2-9f770d76138d",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "85FDOezVftdV",
    "outputId": "2161c5ec-a6ef-4d23-cbdb-c738ef5e53a3",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "geneID                                                                      \n",
      "ENSG00000000003  122.089924        0.074891  0.407199  0.183919  0.854077   \n",
      "ENSG00000000419  186.335951       -0.068959  0.340854 -0.202312  0.839673   \n",
      "ENSG00000000457   22.572713        0.318489  0.877325  0.363023  0.716588   \n",
      "ENSG00000000460   62.835468       -0.390557  0.546729 -0.714351  0.475010   \n",
      "ENSG00000001036  223.518204        0.260500  0.317676  0.820018  0.412206   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSG00000310385   36.231619       -0.225961  0.708885 -0.318756  0.749911   \n",
      "ENSG00000310431   10.435667       -1.356111  1.371015 -0.989129  0.322600   \n",
      "ENSG00000310441   30.178809        0.741581  0.764590  0.969907  0.332093   \n",
      "ENSG00000310517  231.406618        0.288461  0.316413  0.911661  0.361947   \n",
      "ENSG00000310521   16.187346        0.412209  1.046387  0.393935  0.693629   \n",
      "\n",
      "                    padj    Symbol  \n",
      "geneID                              \n",
      "ENSG00000000003  0.99994    TSPAN6  \n",
      "ENSG00000000419  0.99994      DPM1  \n",
      "ENSG00000000457  0.99994     SCYL3  \n",
      "ENSG00000000460  0.99994  C1orf112  \n",
      "ENSG00000001036  0.99994     FUCA2  \n",
      "...                  ...       ...  \n",
      "ENSG00000310385  0.99994       NaN  \n",
      "ENSG00000310431  0.99994       NaN  \n",
      "ENSG00000310441  0.99994       NaN  \n",
      "ENSG00000310517  0.99994       NaN  \n",
      "ENSG00000310521  0.99994       NaN  \n",
      "\n",
      "[10806 rows x 7 columns]\n",
      "Empty DataFrame\n",
      "Columns: [baseMean, log2FoldChange, lfcSE, stat, pvalue, padj, Symbol]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "# 2P\n",
    "twoP = individual_results['2P']\n",
    "twoP = twoP[twoP.baseMean >= 10]\n",
    "print(twoP)\n",
    "\n",
    "psigs = twoP[(twoP['padj'] < padj) & (abs(twoP['log2FoldChange']) > logfold2)]\n",
    "print(psigs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "86ddfbe5-2420-4206-aa34-7efd19602a2d",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "RhWPAkrgfvMO",
    "outputId": "c244ebd2-da43-4c6f-ee86-569d9e156dc4"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "geneID                                                                      \n",
      "ENSG00000000003  122.089924       -0.361927  0.408642 -0.885683  0.375788   \n",
      "ENSG00000000419  186.335951       -0.753888  0.345216 -2.183818  0.028976   \n",
      "ENSG00000000457   22.572713       -0.343233  0.889108 -0.386042  0.699466   \n",
      "ENSG00000000460   62.835468       -0.181457  0.535458 -0.338883  0.734698   \n",
      "ENSG00000001036  223.518204       -0.307100  0.319768 -0.960383  0.336862   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSG00000310385   36.231619       -0.198024  0.698148 -0.283641  0.776685   \n",
      "ENSG00000310431   10.435667       -1.958939  1.396087 -1.403164  0.160568   \n",
      "ENSG00000310441   30.178809        0.434954  0.763870  0.569408  0.569079   \n",
      "ENSG00000310517  231.406618        0.106785  0.315092  0.338900  0.734685   \n",
      "ENSG00000310521   16.187346       -0.120230  1.055230 -0.113937  0.909287   \n",
      "\n",
      "                     padj    Symbol  \n",
      "geneID                               \n",
      "ENSG00000000003  0.999357    TSPAN6  \n",
      "ENSG00000000419  0.999357      DPM1  \n",
      "ENSG00000000457  0.999357     SCYL3  \n",
      "ENSG00000000460  0.999357  C1orf112  \n",
      "ENSG00000001036  0.999357     FUCA2  \n",
      "...                   ...       ...  \n",
      "ENSG00000310385  0.999357       NaN  \n",
      "ENSG00000310431  0.999357       NaN  \n",
      "ENSG00000310441  0.999357       NaN  \n",
      "ENSG00000310517  0.999357       NaN  \n",
      "ENSG00000310521  0.999357       NaN  \n",
      "\n",
      "[10806 rows x 7 columns]\n",
      "Empty DataFrame\n",
      "Columns: [baseMean, log2FoldChange, lfcSE, stat, pvalue, padj, Symbol]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "# 3L\n",
    "threeL = individual_results['3L']\n",
    "threeL = threeL[threeL.baseMean >= 10]\n",
    "print(threeL)\n",
    "\n",
    "lsigs = threeL[(threeL['padj'] < padj) & (abs(threeL['log2FoldChange']) > logfold2)]\n",
    "print(lsigs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "9828d07c-5c7c-45b1-a412-c1ef3bd66194",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "vmHIC6ahXoD7",
    "outputId": "5de00066-c5ba-4167-9672-99b10b7c5868"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
      "geneID                                                                      \n",
      "ENSG00000000003  122.089924        0.107656  0.410745  0.262098  0.793246   \n",
      "ENSG00000000419  186.335951        0.012525  0.343000  0.036516  0.970871   \n",
      "ENSG00000000457   22.572713        0.690554  0.873803  0.790285  0.429361   \n",
      "ENSG00000000460   62.835468       -0.129523  0.546729 -0.236906  0.812730   \n",
      "ENSG00000001036  223.518204        0.045800  0.322433  0.142045  0.887045   \n",
      "...                     ...             ...       ...       ...       ...   \n",
      "ENSG00000310385   36.231619       -0.012234  0.710462 -0.017220  0.986261   \n",
      "ENSG00000310431   10.435667       -0.510116  1.328175 -0.384073  0.700925   \n",
      "ENSG00000310441   30.178809       -0.175723  0.803579 -0.218675  0.826903   \n",
      "ENSG00000310517  231.406618       -0.032686  0.322664 -0.101299  0.919313   \n",
      "ENSG00000310521   16.187346        0.580132  1.050523  0.552232  0.580789   \n",
      "\n",
      "                     padj    Symbol  \n",
      "geneID                               \n",
      "ENSG00000000003  0.999728    TSPAN6  \n",
      "ENSG00000000419  0.999728      DPM1  \n",
      "ENSG00000000457  0.999728     SCYL3  \n",
      "ENSG00000000460  0.999728  C1orf112  \n",
      "ENSG00000001036  0.999728     FUCA2  \n",
      "...                   ...       ...  \n",
      "ENSG00000310385  0.999728       NaN  \n",
      "ENSG00000310431  0.999728       NaN  \n",
      "ENSG00000310441  0.999728       NaN  \n",
      "ENSG00000310517  0.999728       NaN  \n",
      "ENSG00000310521  0.999728       NaN  \n",
      "\n",
      "[10806 rows x 7 columns]\n",
      "Empty DataFrame\n",
      "Columns: [baseMean, log2FoldChange, lfcSE, stat, pvalue, padj, Symbol]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "# 4C\n",
    "fourC = individual_results['4C']\n",
    "fourC = fourC[fourC.baseMean >= 10]\n",
    "print(fourC)\n",
    "\n",
    "csigs = fourC[(fourC['padj'] < padj) & (abs(fourC['log2FoldChange']) > logfold2)]\n",
    "print(csigs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99ff3c04-7ce3-4a5c-8555-1a3d405ad3c8",
   "metadata": {},
   "source": [
    "# ChEA3\n",
    "This script uses the ChEA3 API to perform transcription factor network analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d0c78c66-6ce4-47b8-8656-ab7775556414",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished writing results to chea3_output.csv\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "import requests\n",
    "\n",
    "url = \"https://maayanlab.cloud/chea3/api/enrich/\"\n",
    "gene_set = [\"FOXM1\",\"SMAD9\",\"MYC\",\"SMAD3\",\"STAT1\",\"STAT3\"]\n",
    "output_filename = \"chea3_output.csv\"\n",
    "\n",
    "# Sends a JSON query to the ChEA3 API.\n",
    "# https://maayanlab.cloud/chea3/\n",
    "query_data = {\n",
    "    \"query_name\":\"myQuery\", \n",
    "    \"gene_set\":gene_set\n",
    "}\n",
    "result = requests.post(url, json=query_data)\n",
    "assert result.ok, f\"Request failed with error code: {result.status_code}. Check the ChEA3 website to make sure the request was correct.\"\n",
    "\n",
    "with open(output_filename, \"w\") as csvfile:\n",
    "    # ChEA3 produces several tables (objects) in its JSON response.\n",
    "    # we get items from the table with this name:\n",
    "    entry = \"ENCODE--ChIP-seq\"\n",
    "    tfs = result.json()[entry]\n",
    "\n",
    "    # Each item in the tfs is a dict. We make the table header using its keys.\n",
    "    fieldnames = tfs[0].keys()\n",
    "    \n",
    "    writer = csv.DictWriter(csvfile, fieldnames)\n",
    "    writer.writeheader()\n",
    "    for row in tfs:\n",
    "        writer.writerow(row)\n",
    "\n",
    "print(f\"Finished writing results to {output_filename}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "81c74759-106a-40e2-a41f-a75c8d4f0d0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print the result's data (warning: long)\n",
    "# result.json()\n",
    "\n",
    "# Print the result's text (JSON if successful, HTML if failure)\n",
    "# result.text"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a82c9b46-23f2-476f-bc66-e381e91dc599",
   "metadata": {},
   "source": [
    "# Step 6: Graph of transcription factors\n",
    "Generates graphs for each control-treatment pair using an R program written by Randall J. Eck."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a62a9a1b-30c2-42d0-af8e-21654bf23ad8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ahdog/Projects/csci597k_final_project\n",
      "\u001b[?25h\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hsample_ENCODE_ChIP-seq.tsv sample_ENCODE_ChIP-seq.png\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"Reading from sample_ENCODE_ChIP-seq.tsv printing to sample_ENCODE_ChIP-seq.png\"\n",
      "\u001b[?25h\u001b[1mRows: \u001b[22m\u001b[34m118\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m12\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): Query Name, Set_name, TF, Library, Overlapping_Genes\n",
      "\u001b[32mdbl\u001b[39m (7): Rank, Scaled Rank, Intersect, Set length, FET p-value, FDR, Odds Ratio\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hWarning message:\n",
      "\u001b[1m\u001b[22mRemoved 1 row containing missing values or values outside the scale range\n",
      "(`geom_point()`). \n",
      "\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "!pwd\n",
    "\n",
    "!Rscript ./transcriptionFactors.R sample_ENCODE_ChIP-seq.tsv sample_ENCODE_ChIP-seq.png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42f21350-c8b2-46a2-bdd8-4e02fa1f3466",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
