{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "WM63-41T8O4g"
   },
   "source": [
    "# Overview\n",
    "\n",
    "Notebook by Sky Duryee. Code authors: Ryan Critchlow, Logan Day, Theo Hytopoulos, Vivian White, Aidan Clements, Siddhu Bhimireddy, Randall J. Eck, Sky Duryee, and possibly more.\n",
    "\n",
    "This notebook has most of the code we used to perform RNA-seq in CSCI597k for group 4: transcription factor network analysis. (Some is included in external files.)\n",
    "\n",
    "It takes a little bit of elbow grease to get this working on anything other than our microglia samples. **Unfortunately, some of the scripts rely on the files having specific names like \"01 Agrin Coating\" or \"0M_output.csv\".**\n",
    "\n",
    "Process:\n",
    "1. Start: FASTQ files\n",
    "2. RNA STAR\n",
    "3. Sambamba\n",
    "4. htseq count\n",
    "5. DESeq2\n",
    "6. ChEA3\n",
    "7. Graph R script"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "-ZYCixAXCsZy"
   },
   "source": [
    "# Things to install\n",
    "\n",
    "Here are some things you'll need to install. (Probably incomplete.) These were all run on Linux."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Python packages\n",
    "```\n",
    "requests\n",
    "pandas\n",
    "```\n",
    "* R packages\n",
    "```\n",
    "readxl\n",
    "ggplot2\n",
    "dplyr \n",
    "ggrepel\n",
    "readr\n",
    "```\n",
    "* Bioconductor R packages (These should be installed automatically by the scripts themselves)\n",
    "```\n",
    "DESeq2\n",
    "org.Hs.eg.db\n",
    "```\n",
    "* packages from package manager (e.g. apt)\n",
    "```\n",
    "sambamba\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "I6x79ATaCuXL"
   },
   "outputs": [],
   "source": [
    "!sudo apt install sambamba"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "WmmM86rw9OHa"
   },
   "source": [
    "# RNA STAR\n",
    "\n",
    "We used RNA star to convert the .fastq files into aligned .sam files. We used the human genome gtf file provided at <https://www.ensembl.org/Homo_sapiens/Info/Index?redirect=no>. Alignment takes a long time so we did it using a cluster to speed things up. I was not involved in setting this up. HTCondor was used to manage the cluster. I heard another student mention that they were planning to leave the RNA STAR code on the cluster since it might be needed in the future.\n",
    "\n",
    "Note: I don't have access to the cluster so I tried to restructure the data directory as I think it is on the cluster.\n",
    "\n",
    "Input dir: /cluster/academic/projects/rna-star/STAR/data (not 100% sure)\n",
    "Output dir: results\n",
    "\n",
    "Authors: Theo Hytopoulos, Vivian White, Aidan Clements, possibly 1-2 other folks I didn't record.\n",
    "\n",
    "**Code directory: STAR**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "KOIW_-Lc-YrD"
   },
   "source": [
    "# Sambamba\n",
    "\n",
    "This uses Sambamba to turn the .sam files into .bam files. This is supposed to make htseq count faster.\n",
    "\n",
    "Input dir: results\n",
    "Output dir: sorted_bam\n",
    "\n",
    "Author: Logan Day."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "sKT-I9xfB8GW"
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for dir in results/*/; do\n",
    "  echo dir $dir\n",
    "  if [ -d \"$dir\" ]; then\n",
    "    for subdir in \"$dir\"/*/; do\n",
    "      subdir_name=$(basename \"$subdir\")\n",
    "      if [ -f \"$subdir/Aligned.out.sam\" ]; then\n",
    "        sam_file=\"$subdir/Aligned.out.sam\"\n",
    "        group=\"${dir:9:3}\"\n",
    "        short=\"${subdir_name:0:12}\"\n",
    "        output=$group-$short\n",
    "        ./sambamba view -S --format=bam $sam_file | ./sambamba sort -n -o sorted_bam/${output}.bam /dev/stdin\n",
    "      fi\n",
    "    done\n",
    "  fi\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "K8t39TgkCDx5"
   },
   "source": [
    "# htseq count\n",
    "\n",
    "htseq count is used to create a count matrix for each group of samples, and save it in CSV files. The files are named \"0M_output.csv\" for 00: Matrigel coating, \"1a_output.csv\" for 01: agrin, etc. Unfortunately it will break for differently-named files.\n",
    "\n",
    "Input dir: sorted_bam\n",
    "Output dir: htseqModResults\n",
    "\n",
    "Authors: probably Vivian White, Siddhu Bhimireddy, and Logan Day."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "lhAjpnECC6Ws"
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#!/bin/bash\n",
    "prefixes=(\"0M\" \"0P\" \"1a\" \"2P\" \"3L\" \"4C\")\n",
    "for pre in \"${prefixes[@]}\"; do\n",
    "  bam_files=()\n",
    "  for bam in sorted_bam/\"$pre\"*.bam; do\n",
    "    bam_files+=($bam)\n",
    "  done\n",
    "  echo counting ${bam_files[@]}\n",
    "    htseq-count --format=bam --order=name --stranded=yes \"${bam_files[@]}\" Homo_sapiens.GRCh38.113.gtf > htseqModResults/\"$pre\"_output.csv\n",
    "  echo output to htseqModResults/\"$pre\"_output.csv\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "mzVA79fOC7ON"
   },
   "source": [
    "# DESeq2\n",
    "\n",
    "Performs differential gene expression analysis. Filters by p-value and log2FoldChange as well. See the comment in the R code for info on adjusting filters.\n",
    "\n",
    "I was not able to get this file to run. I got an error: `None of the keys entered are valid keys for 'ENSEMBL'.` The author, Ryan Critchlow, insists that it works on his machine.\n",
    "\n",
    "Input dir: htseqModResults\n",
    "Output dir: filteredResults\n",
    "\n",
    "Author: Ryan Critchlow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "UxCYghW7MviD"
   },
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "FFk0leU0MJ3Q",
    "outputId": "b0f33a4d-3559-445c-adb8-14805c4eed40"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* installing *source* package ‘DESeq2’ ...\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/Rcpp/include' -I'/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/RcppArmadillo/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-4A2Reg/r-base-4.1.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c DESeq2.cpp -o DESeq2.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/Rcpp/include' -I'/home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/RcppArmadillo/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-4A2Reg/r-base-4.1.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c RcppExports.cpp -o RcppExports.o\n",
      "g++ -std=gnu++14 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -o DESeq2.so DESeq2.o RcppExports.o -llapack -lblas -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/ahdog/R/x86_64-pc-linux-gnu-library/4.1/00LOCK-DESeq2/00new/DESeq2/libs\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (DESeq2)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "Bioconductor version 3.14 (BiocManager 1.30.25), R 4.1.2 (2021-11-01)\n",
      "Installing package(s) 'DESeq2'\n",
      "Warning in install.packages(...) : 'lib = \"packages\"' is not writable\n",
      "trying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/DESeq2_1.34.0.tar.gz'\n",
      "Content type 'application/octet-stream' length 2088431 bytes (2.0 MB)\n",
      "==================================================\n",
      "downloaded 2.0 MB\n",
      "\n",
      "\n",
      "The downloaded source packages are in\n",
      "\t‘/tmp/RtmpMMOpO0/downloaded_packages’\n",
      "converting counts to integer mode\n",
      "estimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates\n",
      "mean-dispersion relationship\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "final dispersion estimates\n",
      "fitting model and testing\n",
      "'select()' returned 1:1 mapping between keys and columns\n",
      "Error in .testForValidKeys(x, keys, keytype, fks) : \n",
      "  None of the keys entered are valid keys for 'ENSEMBL'. Please use the keys method to see a listing of valid arguments.\n"
     ]
    },
    {
     "ename": "RInterpreterError",
     "evalue": "Failed to parse and evaluate line '\\n.libPaths(\".\")\\n\\n# Install BiocManager (if not already installed)\\nif (!requireNamespace(\"BiocManager\", quietly = TRUE))\\n  install.packages(\"BiocManager\")\\n\\n# Install DESeq2 from Bioconductor\\nBiocManager::install(\"DESeq2\", force = TRUE, lib = \"packages\")\\n\\n# Load DESeq2 library\\nlibrary(DESeq2)\\n\\n# used for count matrix\\nlibrary(dplyr)\\n\\nif (!requireNamespace(\"org.Hs.eg.db\", quietly = TRUE)) {\\n  BiocManager::install(\"org.Hs.eg.db\", lib = \"packages\", force = TRUE)\\n}\\nlibrary(org.Hs.eg.db)\\n\\n# Function to map Ensembl gene IDs to gene symbols\\nadd_gene_symbols <- function(data) {\\n  gene_ids <- rownames(data)\\n  annots <- select(org.Hs.eg.db,\\n                   keys = gene_ids,\\n                   columns = \"SYMBOL\",\\n                   keytype = \"ENSEMBL\")\\n  data <- merge(data,\\n                annots,\\n                by.x = \"row.names\",\\n                by.y = \"ENSEMBL\",\\n                all.x = TRUE)\\n  rownames(data) <- data$Row.names\\n  data <- data[, -1]  # Remove redundant Row.names column\\n  return(data)\\n}\\n\\nrun_individual_tests <- function(count_matrix,\\n                                 treatments,\\n                                 colData,\\n                                 control_label_0M,\\n                                 control_label_0P) {\\n  # Ensure condition is a factor\\n  colData$condition <- as.factor(colData$condition)\\n\\n  # Initialize DESeq2 dataset\\n  dds <- DESeqDataSetFromMatrix(countData = count_matrix,\\n                                colData = colData,\\n                                design = ~ condition)\\n\\n  # filter out rows with low gene counts\\n  dds <- dds[rowSums(counts(dds)) > 1, ]\\n\\n  # Run DESeq2\\n  dds <- DESeq(dds)\\n\\n  # Store results for each treatment\\n  results_list <- list()\\n\\n  # Filter function without additional libraries\\n  filter_results <- function(res_df,\\n                             lfc_thresh = 1,\\n                             padj_thresh = 0.05) {\\n    # Remove rows with NA values in log2FoldChange or padj\\n    res_df <- res_df[!is.na(res_df$log2FoldChange) &\\n                       !is.na(res_df$padj), ]\\n    # filter by threshold values\\n    res_df <- res_df[abs(res_df$log2FoldChange) > lfc_thresh &\\n                       res_df$padj < padj_thresh, ]\\n    return(res_df)\\n  }\\n\\n  # Run comparisons for treatments vs. 0M control\\n  for (treatment in treatments) {\\n    res <- results(dds, contrast = c(\"condition\", treatment, control_label_0M))\\n    res_df <- as.data.frame(res)  # Convert to data frame\\n    filtered_res <- filter_results(res_df)  # Filter results\\n    # Add gene symbols\\n    filtered_res <- add_gene_symbols(filtered_res)\\n    results_list[[paste0(treatment, \"_vs_\", control_label_0M)]] <- filtered_res\\n  }\\n\\n  # Run comparisons for treatments vs. 0P control\\n  for (treatment in treatments) {\\n    res <- results(dds, contrast = c(\"condition\", treatment, control_label_0P))\\n    res_df <- as.data.frame(res)  # Convert to data frame\\n    filtered_res <- filter_results(res_df)  # Filter results\\n    # Add gene symbols\\n    filtered_res <- add_gene_symbols(filtered_res)\\n    results_list[[paste0(treatment, \"_vs_\", control_label_0P)]] <- filtered_res\\n  }\\n\\n  return(results_list)\\n}\\n\\n# Define the function\\nload_count_matrix <- function() {\\n  # File paths for the counts files\\n  counts_files <- list(\\n    \"0M\" = \"htseqModResults/0M_output.csv\",\\n    \"0P\" = \"htseqModResults/0P_output.csv\",\\n    \"1A\" = \"htseqModResults/1a_output.csv\",\\n    \"2P\" = \"htseqModResults/2P_output.csv\",\\n    \"3L\" = \"htseqModResults/3L_output.csv\",\\n    \"4C\" = \"htseqModResults/4C_output.csv\"\\n  )\\n\\n  # Read all files into a list of data frames\\n  dataframes <- lapply(names(counts_files), function(sample) {\\n    file_path <- counts_files[[sample]]\\n    df <- read.csv(\\n      file_path,\\n      sep = \"\\\\t\",\\n      header = FALSE,\\n      stringsAsFactors = FALSE\\n    )\\n    colnames(df) <- c(\"geneID\", paste0(sample, \"_col\", seq(1, ncol(df) - 1)))\\n    return(df)\\n  })\\n\\n  # Merge all data frames on \"geneID\" using Reduce and full_join\\n  count_matrix <- Reduce(function(x, y) {\\n    full_join(x, y, by = \"geneID\")\\n  }, dataframes)\\n\\n  # Replace NA values with zeros\\n  count_matrix[is.na(count_matrix)] <- 0\\n\\n  # Filter out rows (genes) with zero counts across all samples\\n  count_matrix <- count_matrix[rowSums(count_matrix[, -1]) > 0, ]\\n\\n  # Filter gene IDs that start with \\'E\\'\\n  count_matrix <- count_matrix[grepl(\"^E\", count_matrix$geneID), ]\\n\\n  # Set geneID as rownames and remove the geneID column\\n  rownames(count_matrix) <- count_matrix$geneID\\n  count_matrix <- count_matrix[, -1]  # Drop the geneID column\\n\\n  return(count_matrix)\\n}\\n\\n# obtain count matrix\\ncount_matrix <- load_count_matrix()\\n\\ncount_matrix <- round(count_matrix)  # Ensure counts are integers\\n\\n# Define metadata\\ncolData <- data.frame(\\n  condition = c(\\n    rep(\"0M\", 8),\\n    rep(\"0P\", 8),\\n    rep(\"1A\", 6),\\n    rep(\"2P\", 8),\\n    rep(\"3L\", 8),\\n    rep(\"4C\", 8)\\n  ),\\n  row.names = colnames(count_matrix)\\n)\\n\\nall(colnames(count_matrix) %in% rownames (colData))\\n\\nall(colnames(count_matrix) == rownames(colData))\\n\\n# Define treatments\\ntreatments <- c(\"1A\", \"2P\", \"3L\", \"4C\")\\n\\noutput_dir <- \"filteredResults\"\\n\\n# run the tests\\nresults <- run_individual_tests(\\n  count_matrix = count_matrix,\\n  treatments = treatments,\\n  colData = colData,\\n  control_label_0M = \"0M\",\\n  control_label_0P = \"0P\"\\n)\\n\\n# Inspect results and print to files\\nfor (name in names(results)) {\\n  cat(\"\\\\nComparison:\", name, \"\\\\n\")\\n  print(results[[name]])\\n  # Write filtered results to CSV in the downloads directory\\n  write.csv(results[[name]],\\n            file = file.path(output_dir\\n                             , paste0(name, \"_filtered.csv\")),\\n            row.names = TRUE)\\n}\\n'.\nR error message: \"Error in .testForValidKeys(x, keys, keytype, fks) : \\n  None of the keys entered are valid keys for 'ENSEMBL'. Please use the keys method to see a listing of valid arguments.\"\nR stdout:\n'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n'help(\"repositories\", package = \"BiocManager\")' for details.\nReplacement repositories:\n    CRAN: https://cloud.r-project.org\nBioconductor version 3.14 (BiocManager 1.30.25), R 4.1.2 (2021-11-01)\nInstalling package(s) 'DESeq2'\nWarning in install.packages(...) : 'lib = \"packages\"' is not writable\ntrying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/DESeq2_1.34.0.tar.gz'\nContent type 'application/octet-stream' length 2088431 bytes (2.0 MB)\n==================================================\ndownloaded 2.0 MB\n\n\nThe downloaded source packages are in\n\t‘/tmp/RtmpMMOpO0/downloaded_packages’\nconverting counts to integer mode\nestimating size factors\nestimating dispersions\ngene-wise dispersion estimates\nmean-dispersion relationship\n-- note: fitType='parametric', but the dispersion trend was not well captured by the\n   function: y = a/x + b, and a local regression fit was automatically substituted.\n   specify fitType='local' or 'mean' to avoid this message next time.\nfinal dispersion estimates\nfitting model and testing\n'select()' returned 1:1 mapping between keys and columns\nError in .testForValidKeys(x, keys, keytype, fks) : \n  None of the keys entered are valid keys for 'ENSEMBL'. Please use the keys method to see a listing of valid arguments.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/ipython/rmagic.py\u001b[0m in \u001b[0;36meval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    406\u001b[0m                 \u001b[0mr_expr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mri\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 407\u001b[0;31m                 value, visible = ri.evalr_expr_with_visible(\n\u001b[0m\u001b[1;32m    408\u001b[0m                     \u001b[0mr_expr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/rinterface.py\u001b[0m in \u001b[0;36mevalr_expr_with_visible\u001b[0;34m(expr, envir)\u001b[0m\n\u001b[1;32m    193\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0merror_occured\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 194\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0membedded\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mRRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_rinterface\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_geterrmessage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    195\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconversion\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cdata_to_rinterface\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr_res\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error in .testForValidKeys(x, keys, keytype, fks) : \n  None of the keys entered are valid keys for 'ENSEMBL'. Please use the keys method to see a listing of valid arguments.\n",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mRInterpreterError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_27783/2337799079.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'R'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'\\n.libPaths(\".\")\\n\\n# Install BiocManager (if not already installed)\\nif (!requireNamespace(\"BiocManager\", quietly = TRUE))\\n  install.packages(\"BiocManager\")\\n\\n# Install DESeq2 from Bioconductor\\nBiocManager::install(\"DESeq2\", force = TRUE, lib = \"packages\")\\n\\n# Load DESeq2 library\\nlibrary(DESeq2)\\n\\n# used for count matrix\\nlibrary(dplyr)\\n\\nif (!requireNamespace(\"org.Hs.eg.db\", quietly = TRUE)) {\\n  BiocManager::install(\"org.Hs.eg.db\", lib = \"packages\", force = TRUE)\\n}\\nlibrary(org.Hs.eg.db)\\n\\n# Function to map Ensembl gene IDs to gene symbols\\nadd_gene_symbols <- function(data) {\\n  gene_ids <- rownames(data)\\n  annots <- select(org.Hs.eg.db,\\n                   keys = gene_ids,\\n                   columns = \"SYMBOL\",\\n                   keytype = \"ENSEMBL\")\\n  data <- merge(data,\\n                annots,\\n                by.x = \"row.names\",\\n                by.y = \"ENSEMBL\",\\n                all.x = TRUE)\\n  rownames(data) <- data$Row.names\\n  data <- data[, -1]  # Remove redundant Row.names column\\n  return(data)\\n}\\n\\nrun_individual_tests <- function(count_matrix,\\n                                 treatments,\\n                                 colData,\\n                                 control_label_0M,\\n                                 control_label_0P) {\\n  # Ensure condition is a factor\\n  colData$condition <- as.factor(colData$condition)\\n\\n  # Initialize DESeq2 dataset\\n  dds <- DESeqDataSetFromMatrix(countData = count_matrix,\\n                                colData = colData,\\n                                design = ~ condition)\\n\\n  # filter out rows with low gene counts\\n  dds <- dds[rowSums(counts(dds)) > 1, ]\\n\\n  # Run DESeq2\\n  dds <- DESeq(dds)\\n\\n  # Store results for each treatment\\n  results_list <- list()\\n\\n  # Filter function without additional libraries\\n  filter_results <- function(res_df,\\n                             lfc_thresh = 1,\\n                             padj_thresh = 0.05) {\\n    # Remove rows with NA values in log2FoldChange or padj\\n    res_df <- res_df[!is.na(res_df$log2FoldChange) &\\n                       !is.na(res_df$padj), ]\\n    # filter by threshold values\\n    res_df <- res_df[abs(res_df$log2FoldChange) > lfc_thresh &\\n                       res_df$padj < padj_thresh, ]\\n    return(res_df)\\n  }\\n\\n  # Run comparisons for treatments vs. 0M control\\n  for (treatment in treatments) {\\n    res <- results(dds, contrast = c(\"condition\", treatment, control_label_0M))\\n    res_df <- as.data.frame(res)  # Convert to data frame\\n    filtered_res <- filter_results(res_df)  # Filter results\\n    # Add gene symbols\\n    filtered_res <- add_gene_symbols(filtered_res)\\n    results_list[[paste0(treatment, \"_vs_\", control_label_0M)]] <- filtered_res\\n  }\\n\\n  # Run comparisons for treatments vs. 0P control\\n  for (treatment in treatments) {\\n    res <- results(dds, contrast = c(\"condition\", treatment, control_label_0P))\\n    res_df <- as.data.frame(res)  # Convert to data frame\\n    filtered_res <- filter_results(res_df)  # Filter results\\n    # Add gene symbols\\n    filtered_res <- add_gene_symbols(filtered_res)\\n    results_list[[paste0(treatment, \"_vs_\", control_label_0P)]] <- filtered_res\\n  }\\n\\n  return(results_list)\\n}\\n\\n# Define the function\\nload_count_matrix <- function() {\\n  # File paths for the counts files\\n  counts_files <- list(\\n    \"0M\" = \"htseqModResults/0M_output.csv\",\\n    \"0P\" = \"htseqModResults/0P_output.csv\",\\n    \"1A\" = \"htseqModResults/1a_output.csv\",\\n    \"2P\" = \"htseqModResults/2P_output.csv\",\\n    \"3L\" = \"htseqModResults/3L_output.csv\",\\n    \"4C\" = \"htseqModResults/4C_output.csv\"\\n  )\\n\\n  # Read all files into a list of data frames\\n  dataframes <- lapply(names(counts_files), function(sample) {\\n    file_path <- counts_files[[sample]]\\n    df <- read.csv(\\n      file_path,\\n      sep = \"\\\\t\",\\n      header = FALSE,\\n      stringsAsFactors = FALSE\\n    )\\n    colnames(df) <- c(\"geneID\", paste0(sample, \"_col\", seq(1, ncol(df) - 1)))\\n    return(df)\\n  })\\n\\n  # Merge all data frames on \"geneID\" using Reduce and full_join\\n  count_matrix <- Reduce(function(x, y) {\\n    full_join(x, y, by = \"geneID\")\\n  }, dataframes)\\n\\n  # Replace NA values with zeros\\n  count_matrix[is.na(count_matrix)] <- 0\\n\\n  # Filter out rows (genes) with zero counts across all samples\\n  count_matrix <- count_matrix[rowSums(count_matrix[, -1]) > 0, ]\\n\\n  # Filter gene IDs that start with \\'E\\'\\n  count_matrix <- count_matrix[grepl(\"^E\", count_matrix$geneID), ]\\n\\n  # Set geneID as rownames and remove the geneID column\\n  rownames(count_matrix) <- count_matrix$geneID\\n  count_matrix <- count_matrix[, -1]  # Drop the geneID column\\n\\n  return(count_matrix)\\n}\\n\\n# obtain count matrix\\ncount_matrix <- load_count_matrix()\\n\\ncount_matrix <- round(count_matrix)  # Ensure counts are integers\\n\\n# Define metadata\\ncolData <- data.frame(\\n  condition = c(\\n    rep(\"0M\", 8),\\n    rep(\"0P\", 8),\\n    rep(\"1A\", 6),\\n    rep(\"2P\", 8),\\n    rep(\"3L\", 8),\\n    rep(\"4C\", 8)\\n  ),\\n  row.names = colnames(count_matrix)\\n)\\n\\nall(colnames(count_matrix) %in% rownames (colData))\\n\\nall(colnames(count_matrix) == rownames(colData))\\n\\n# Define treatments\\ntreatments <- c(\"1A\", \"2P\", \"3L\", \"4C\")\\n\\noutput_dir <- \"filteredResults\"\\n\\n# run the tests\\nresults <- run_individual_tests(\\n  count_matrix = count_matrix,\\n  treatments = treatments,\\n  colData = colData,\\n  control_label_0M = \"0M\",\\n  control_label_0P = \"0P\"\\n)\\n\\n# Inspect results and print to files\\nfor (name in names(results)) {\\n  cat(\"\\\\nComparison:\", name, \"\\\\n\")\\n  print(results[[name]])\\n  # Write filtered results to CSV in the downloads directory\\n  write.csv(results[[name]],\\n            file = file.path(output_dir\\n                             , paste0(name, \"_filtered.csv\")),\\n            row.names = TRUE)\\n}\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/usr/lib/python3/dist-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2417\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2418\u001b[0m                 \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2419\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2420\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2421\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<decorator-gen-121>\u001b[0m in \u001b[0;36mR\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3/dist-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/ipython/rmagic.py\u001b[0m in \u001b[0;36mR\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    982\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mendswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    983\u001b[0m                 \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 984\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    985\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    986\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdevice\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mDEVICES_STATIC\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/ipython/rmagic.py\u001b[0m in \u001b[0;36mR\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    947\u001b[0m                     \u001b[0mreturn_output\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    948\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 949\u001b[0;31m                 \u001b[0mtext_result\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvisible\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0meval\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    950\u001b[0m                 \u001b[0mtext_output\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mtext_result\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    951\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mvisible\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/ipython/rmagic.py\u001b[0m in \u001b[0;36meval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    411\u001b[0m                 \u001b[0;31m# Otherwise next return seems to have copy of error.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    412\u001b[0m                 \u001b[0mwarning_or_other_msg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 413\u001b[0;31m                 raise RInterpreterError(code, str(exception),\n\u001b[0m\u001b[1;32m    414\u001b[0m                                         warning_or_other_msg)\n\u001b[1;32m    415\u001b[0m             \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRInterpreterError\u001b[0m: Failed to parse and evaluate line '\\n.libPaths(\".\")\\n\\n# Install BiocManager (if not already installed)\\nif (!requireNamespace(\"BiocManager\", quietly = TRUE))\\n  install.packages(\"BiocManager\")\\n\\n# Install DESeq2 from Bioconductor\\nBiocManager::install(\"DESeq2\", force = TRUE, lib = \"packages\")\\n\\n# Load DESeq2 library\\nlibrary(DESeq2)\\n\\n# used for count matrix\\nlibrary(dplyr)\\n\\nif (!requireNamespace(\"org.Hs.eg.db\", quietly = TRUE)) {\\n  BiocManager::install(\"org.Hs.eg.db\", lib = \"packages\", force = TRUE)\\n}\\nlibrary(org.Hs.eg.db)\\n\\n# Function to map Ensembl gene IDs to gene symbols\\nadd_gene_symbols <- function(data) {\\n  gene_ids <- rownames(data)\\n  annots <- select(org.Hs.eg.db,\\n                   keys = gene_ids,\\n                   columns = \"SYMBOL\",\\n                   keytype = \"ENSEMBL\")\\n  data <- merge(data,\\n                annots,\\n                by.x = \"row.names\",\\n                by.y = \"ENSEMBL\",\\n                all.x = TRUE)\\n  rownames(data) <- data$Row.names\\n  data <- data[, -1]  # Remove redundant Row.names column\\n  return(data)\\n}\\n\\nrun_individual_tests <- function(count_matrix,\\n                                 treatments,\\n                                 colData,\\n                                 control_label_0M,\\n                                 control_label_0P) {\\n  # Ensure condition is a factor\\n  colData$condition <- as.factor(colData$condition)\\n\\n  # Initialize DESeq2 dataset\\n  dds <- DESeqDataSetFromMatrix(countData = count_matrix,\\n                                colData = colData,\\n                                design = ~ condition)\\n\\n  # filter out rows with low gene counts\\n  dds <- dds[rowSums(counts(dds)) > 1, ]\\n\\n  # Run DESeq2\\n  dds <- DESeq(dds)\\n\\n  # Store results for each treatment\\n  results_list <- list()\\n\\n  # Filter function without additional libraries\\n  filter_results <- function(res_df,\\n                             lfc_thresh = 1,\\n                             padj_thresh = 0.05) {\\n    # Remove rows with NA values in log2FoldChange or padj\\n    res_df <- res_df[!is.na(res_df$log2FoldChange) &\\n                       !is.na(res_df$padj), ]\\n    # filter by threshold values\\n    res_df <- res_df[abs(res_df$log2FoldChange) > lfc_thresh &\\n                       res_df$padj < padj_thresh, ]\\n    return(res_df)\\n  }\\n\\n  # Run comparisons for treatments vs. 0M control\\n  for (treatment in treatments) {\\n    res <- results(dds, contrast = c(\"condition\", treatment, control_label_0M))\\n    res_df <- as.data.frame(res)  # Convert to data frame\\n    filtered_res <- filter_results(res_df)  # Filter results\\n    # Add gene symbols\\n    filtered_res <- add_gene_symbols(filtered_res)\\n    results_list[[paste0(treatment, \"_vs_\", control_label_0M)]] <- filtered_res\\n  }\\n\\n  # Run comparisons for treatments vs. 0P control\\n  for (treatment in treatments) {\\n    res <- results(dds, contrast = c(\"condition\", treatment, control_label_0P))\\n    res_df <- as.data.frame(res)  # Convert to data frame\\n    filtered_res <- filter_results(res_df)  # Filter results\\n    # Add gene symbols\\n    filtered_res <- add_gene_symbols(filtered_res)\\n    results_list[[paste0(treatment, \"_vs_\", control_label_0P)]] <- filtered_res\\n  }\\n\\n  return(results_list)\\n}\\n\\n# Define the function\\nload_count_matrix <- function() {\\n  # File paths for the counts files\\n  counts_files <- list(\\n    \"0M\" = \"htseqModResults/0M_output.csv\",\\n    \"0P\" = \"htseqModResults/0P_output.csv\",\\n    \"1A\" = \"htseqModResults/1a_output.csv\",\\n    \"2P\" = \"htseqModResults/2P_output.csv\",\\n    \"3L\" = \"htseqModResults/3L_output.csv\",\\n    \"4C\" = \"htseqModResults/4C_output.csv\"\\n  )\\n\\n  # Read all files into a list of data frames\\n  dataframes <- lapply(names(counts_files), function(sample) {\\n    file_path <- counts_files[[sample]]\\n    df <- read.csv(\\n      file_path,\\n      sep = \"\\\\t\",\\n      header = FALSE,\\n      stringsAsFactors = FALSE\\n    )\\n    colnames(df) <- c(\"geneID\", paste0(sample, \"_col\", seq(1, ncol(df) - 1)))\\n    return(df)\\n  })\\n\\n  # Merge all data frames on \"geneID\" using Reduce and full_join\\n  count_matrix <- Reduce(function(x, y) {\\n    full_join(x, y, by = \"geneID\")\\n  }, dataframes)\\n\\n  # Replace NA values with zeros\\n  count_matrix[is.na(count_matrix)] <- 0\\n\\n  # Filter out rows (genes) with zero counts across all samples\\n  count_matrix <- count_matrix[rowSums(count_matrix[, -1]) > 0, ]\\n\\n  # Filter gene IDs that start with \\'E\\'\\n  count_matrix <- count_matrix[grepl(\"^E\", count_matrix$geneID), ]\\n\\n  # Set geneID as rownames and remove the geneID column\\n  rownames(count_matrix) <- count_matrix$geneID\\n  count_matrix <- count_matrix[, -1]  # Drop the geneID column\\n\\n  return(count_matrix)\\n}\\n\\n# obtain count matrix\\ncount_matrix <- load_count_matrix()\\n\\ncount_matrix <- round(count_matrix)  # Ensure counts are integers\\n\\n# Define metadata\\ncolData <- data.frame(\\n  condition = c(\\n    rep(\"0M\", 8),\\n    rep(\"0P\", 8),\\n    rep(\"1A\", 6),\\n    rep(\"2P\", 8),\\n    rep(\"3L\", 8),\\n    rep(\"4C\", 8)\\n  ),\\n  row.names = colnames(count_matrix)\\n)\\n\\nall(colnames(count_matrix) %in% rownames (colData))\\n\\nall(colnames(count_matrix) == rownames(colData))\\n\\n# Define treatments\\ntreatments <- c(\"1A\", \"2P\", \"3L\", \"4C\")\\n\\noutput_dir <- \"filteredResults\"\\n\\n# run the tests\\nresults <- run_individual_tests(\\n  count_matrix = count_matrix,\\n  treatments = treatments,\\n  colData = colData,\\n  control_label_0M = \"0M\",\\n  control_label_0P = \"0P\"\\n)\\n\\n# Inspect results and print to files\\nfor (name in names(results)) {\\n  cat(\"\\\\nComparison:\", name, \"\\\\n\")\\n  print(results[[name]])\\n  # Write filtered results to CSV in the downloads directory\\n  write.csv(results[[name]],\\n            file = file.path(output_dir\\n                             , paste0(name, \"_filtered.csv\")),\\n            row.names = TRUE)\\n}\\n'.\nR error message: \"Error in .testForValidKeys(x, keys, keytype, fks) : \\n  None of the keys entered are valid keys for 'ENSEMBL'. Please use the keys method to see a listing of valid arguments.\"\nR stdout:\n'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n'help(\"repositories\", package = \"BiocManager\")' for details.\nReplacement repositories:\n    CRAN: https://cloud.r-project.org\nBioconductor version 3.14 (BiocManager 1.30.25), R 4.1.2 (2021-11-01)\nInstalling package(s) 'DESeq2'\nWarning in install.packages(...) : 'lib = \"packages\"' is not writable\ntrying URL 'https://bioconductor.org/packages/3.14/bioc/src/contrib/DESeq2_1.34.0.tar.gz'\nContent type 'application/octet-stream' length 2088431 bytes (2.0 MB)\n==================================================\ndownloaded 2.0 MB\n\n\nThe downloaded source packages are in\n\t‘/tmp/RtmpMMOpO0/downloaded_packages’\nconverting counts to integer mode\nestimating size factors\nestimating dispersions\ngene-wise dispersion estimates\nmean-dispersion relationship\n-- note: fitType='parametric', but the dispersion trend was not well captured by the\n   function: y = a/x + b, and a local regression fit was automatically substituted.\n   specify fitType='local' or 'mean' to avoid this message next time.\nfinal dispersion estimates\nfitting model and testing\n'select()' returned 1:1 mapping between keys and columns\nError in .testForValidKeys(x, keys, keytype, fks) : \n  None of the keys entered are valid keys for 'ENSEMBL'. Please use the keys method to see a listing of valid arguments."
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "#####\n",
    "# adjust the lft_thresh (log2FoldChange threshold)\n",
    "# and padj_thresh (adjusted p-value threshold)\n",
    "# in the filter_results function. (You can ctrl+F for this)\n",
    "##### \n",
    "\n",
    ".libPaths(\".\")\n",
    "\n",
    "# Install BiocManager (if not already installed)\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "  install.packages(\"BiocManager\")\n",
    "\n",
    "# Install DESeq2 from Bioconductor\n",
    "BiocManager::install(\"DESeq2\", force = TRUE, lib = \"packages\")\n",
    "\n",
    "# Load DESeq2 library\n",
    "library(DESeq2)\n",
    "\n",
    "# used for count matrix\n",
    "library(dplyr)\n",
    "\n",
    "if (!requireNamespace(\"org.Hs.eg.db\", quietly = TRUE)) {\n",
    "  BiocManager::install(\"org.Hs.eg.db\", lib = \"packages\", force = TRUE)\n",
    "}\n",
    "library(org.Hs.eg.db)\n",
    "\n",
    "# Function to map Ensembl gene IDs to gene symbols\n",
    "add_gene_symbols <- function(data) {\n",
    "  gene_ids <- rownames(data)\n",
    "  annots <- select(org.Hs.eg.db,\n",
    "                   keys = gene_ids,\n",
    "                   columns = \"SYMBOL\",\n",
    "                   keytype = \"ENSEMBL\")\n",
    "  data <- merge(data,\n",
    "                annots,\n",
    "                by.x = \"row.names\",\n",
    "                by.y = \"ENSEMBL\",\n",
    "                all.x = TRUE)\n",
    "  rownames(data) <- data$Row.names\n",
    "  data <- data[, -1]  # Remove redundant Row.names column\n",
    "  return(data)\n",
    "}\n",
    "\n",
    "run_individual_tests <- function(count_matrix,\n",
    "                                 treatments,\n",
    "                                 colData,\n",
    "                                 control_label_0M,\n",
    "                                 control_label_0P) {\n",
    "  # Ensure condition is a factor\n",
    "  colData$condition <- as.factor(colData$condition)\n",
    "\n",
    "  # Initialize DESeq2 dataset\n",
    "  dds <- DESeqDataSetFromMatrix(countData = count_matrix,\n",
    "                                colData = colData,\n",
    "                                design = ~ condition)\n",
    "\n",
    "  # filter out rows with low gene counts\n",
    "  dds <- dds[rowSums(counts(dds)) > 1, ]\n",
    "\n",
    "  # Run DESeq2\n",
    "  dds <- DESeq(dds)\n",
    "\n",
    "  # Store results for each treatment\n",
    "  results_list <- list()\n",
    "\n",
    "  # Filter function without additional libraries\n",
    "  filter_results <- function(res_df,\n",
    "                             lfc_thresh = 1,\n",
    "                             padj_thresh = 0.05) {\n",
    "    # Remove rows with NA values in log2FoldChange or padj\n",
    "    res_df <- res_df[!is.na(res_df$log2FoldChange) &\n",
    "                       !is.na(res_df$padj), ]\n",
    "    # filter by threshold values\n",
    "    res_df <- res_df[abs(res_df$log2FoldChange) > lfc_thresh &\n",
    "                       res_df$padj < padj_thresh, ]\n",
    "    return(res_df)\n",
    "  }\n",
    "\n",
    "  # Run comparisons for treatments vs. 0M control\n",
    "  for (treatment in treatments) {\n",
    "    res <- results(dds, contrast = c(\"condition\", treatment, control_label_0M))\n",
    "    res_df <- as.data.frame(res)  # Convert to data frame\n",
    "    filtered_res <- filter_results(res_df)  # Filter results\n",
    "    # Add gene symbols\n",
    "    filtered_res <- add_gene_symbols(filtered_res)\n",
    "    results_list[[paste0(treatment, \"_vs_\", control_label_0M)]] <- filtered_res\n",
    "  }\n",
    "\n",
    "  # Run comparisons for treatments vs. 0P control\n",
    "  for (treatment in treatments) {\n",
    "    res <- results(dds, contrast = c(\"condition\", treatment, control_label_0P))\n",
    "    res_df <- as.data.frame(res)  # Convert to data frame\n",
    "    filtered_res <- filter_results(res_df)  # Filter results\n",
    "    # Add gene symbols\n",
    "    filtered_res <- add_gene_symbols(filtered_res)\n",
    "    results_list[[paste0(treatment, \"_vs_\", control_label_0P)]] <- filtered_res\n",
    "  }\n",
    "\n",
    "  return(results_list)\n",
    "}\n",
    "\n",
    "# Define the function\n",
    "load_count_matrix <- function() {\n",
    "  # File paths for the counts files\n",
    "  counts_files <- list(\n",
    "    \"0M\" = \"htseqModResults/0M_output.csv\",\n",
    "    \"0P\" = \"htseqModResults/0P_output.csv\",\n",
    "    \"1A\" = \"htseqModResults/1a_output.csv\",\n",
    "    \"2P\" = \"htseqModResults/2P_output.csv\",\n",
    "    \"3L\" = \"htseqModResults/3L_output.csv\",\n",
    "    \"4C\" = \"htseqModResults/4C_output.csv\"\n",
    "  )\n",
    "\n",
    "  # Read all files into a list of data frames\n",
    "  dataframes <- lapply(names(counts_files), function(sample) {\n",
    "    file_path <- counts_files[[sample]]\n",
    "    df <- read.csv(\n",
    "      file_path,\n",
    "      sep = \"\\t\",\n",
    "      header = FALSE,\n",
    "      stringsAsFactors = FALSE\n",
    "    )\n",
    "    colnames(df) <- c(\"geneID\", paste0(sample, \"_col\", seq(1, ncol(df) - 1)))\n",
    "    return(df)\n",
    "  })\n",
    "\n",
    "  # Merge all data frames on \"geneID\" using Reduce and full_join\n",
    "  count_matrix <- Reduce(function(x, y) {\n",
    "    full_join(x, y, by = \"geneID\")\n",
    "  }, dataframes)\n",
    "\n",
    "  # Replace NA values with zeros\n",
    "  count_matrix[is.na(count_matrix)] <- 0\n",
    "\n",
    "  # Filter out rows (genes) with zero counts across all samples\n",
    "  count_matrix <- count_matrix[rowSums(count_matrix[, -1]) > 0, ]\n",
    "\n",
    "  # Filter gene IDs that start with 'E'\n",
    "  count_matrix <- count_matrix[grepl(\"^E\", count_matrix$geneID), ]\n",
    "\n",
    "  # Set geneID as rownames and remove the geneID column\n",
    "  rownames(count_matrix) <- count_matrix$geneID\n",
    "  count_matrix <- count_matrix[, -1]  # Drop the geneID column\n",
    "\n",
    "  return(count_matrix)\n",
    "}\n",
    "\n",
    "# obtain count matrix\n",
    "count_matrix <- load_count_matrix()\n",
    "\n",
    "count_matrix <- round(count_matrix)  # Ensure counts are integers\n",
    "\n",
    "# Define metadata\n",
    "colData <- data.frame(\n",
    "  condition = c(\n",
    "    rep(\"0M\", 8),\n",
    "    rep(\"0P\", 8),\n",
    "    rep(\"1A\", 6),\n",
    "    rep(\"2P\", 8),\n",
    "    rep(\"3L\", 8),\n",
    "    rep(\"4C\", 8)\n",
    "  ),\n",
    "  row.names = colnames(count_matrix)\n",
    ")\n",
    "\n",
    "all(colnames(count_matrix) %in% rownames (colData))\n",
    "\n",
    "all(colnames(count_matrix) == rownames(colData))\n",
    "\n",
    "# Define treatments\n",
    "treatments <- c(\"1A\", \"2P\", \"3L\", \"4C\")\n",
    "\n",
    "output_dir <- \"filteredResults\"\n",
    "\n",
    "# run the tests\n",
    "results <- run_individual_tests(\n",
    "  count_matrix = count_matrix,\n",
    "  treatments = treatments,\n",
    "  colData = colData,\n",
    "  control_label_0M = \"0M\",\n",
    "  control_label_0P = \"0P\"\n",
    ")\n",
    "\n",
    "# Inspect results and print to files\n",
    "for (name in names(results)) {\n",
    "  cat(\"\\nComparison:\", name, \"\\n\")\n",
    "  print(results[[name]])\n",
    "  # Write filtered results to CSV in the downloads directory\n",
    "  write.csv(results[[name]],\n",
    "            file = file.path(output_dir\n",
    "                             , paste0(name, \"_filtered.csv\")),\n",
    "            row.names = TRUE)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering output files (if above script does not work)\n",
    "The above R code filters the files. Since I was not able to get it to run, Ryan used his above code to generate some unfiltered results and sent them to me. (He adjusted the parameters not to filter anything out.) Then I used this script to filter out his results.\n",
    "\n",
    "You can skip running this if you get the above R code to run.\n",
    "\n",
    "Input dir: unfilteredResults (`src_dir` variable)\n",
    "Output dir: pyFilteredResults (`dst_dir` variable)\n",
    "\n",
    "Author: Sky Duryee"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Opened: 1A_vs_0P_filtered.csv\n",
      "Opened: 2P_vs_0M_filtered.csv\n",
      "Opened: 3L_vs_0M_filtered.csv\n",
      "Opened: 4C_vs_0P_filtered.csv\n",
      "Opened: 2P_vs_0P_filtered.csv\n",
      "Opened: 1A_vs_0M_filtered.csv\n",
      "Opened: 4C_vs_0M_filtered.csv\n",
      "Opened: 3L_vs_0P_filtered.csv\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "\n",
    "src_dir = \"unfilteredResults\"\n",
    "dst_dir = \"pyFilteredResults\"\n",
    "\n",
    "padj_thresh = 0.05\n",
    "log2FoldChange_thresh = 0.5\n",
    "\n",
    "Path(dst_dir).mkdir(exist_ok=True)\n",
    "for unfiltered_file in os.listdir(src_dir):\n",
    "    print(f\"Opened: {unfiltered_file}\")\n",
    "    result = pd.read_csv(f\"{src_dir}/{unfiltered_file}\")\n",
    "    # filter using adjust p-value < 0.05 and log2FoldChange > 0.5\n",
    "    result = result[(result[\"padj\"] < padj_thresh) & (abs(result[\"log2FoldChange\"]) > log2FoldChange_thresh)]\n",
    "    result = result.sort_values(\"log2FoldChange\")\n",
    "\n",
    "    # print(result)\n",
    "    result.to_csv(f\"{dst_dir}/{unfiltered_file}\", header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ChEA3\n",
    "Run ChEA3 using genes from the filtered files.\n",
    "\n",
    "This relies on ChEA3's API, so it might not work if their website goes down.\n",
    "\n",
    "The input dir might need to be changed, if you adjust the parameters. See the *DESeq2* and *Filtering Output Files* sections.\n",
    "\n",
    "Input dir: pyFilteredResults (`src_dir` variable) \n",
    "Output dir: chea3Results (`dst_dir` variable)\n",
    "\n",
    "Author: Sky Duryee"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating TSV for pyFilteredResults/1A_vs_0P_filtered.csv...\n",
      "Finished writing results to chea3Results/1A_vs_0P.tsv\n",
      "Generating TSV for pyFilteredResults/2P_vs_0M_filtered.csv...\n",
      "Gene set only has 0 genes. Not producing CSV.\n",
      "Generating TSV for pyFilteredResults/3L_vs_0M_filtered.csv...\n",
      "Finished writing results to chea3Results/3L_vs_0M.tsv\n",
      "Generating TSV for pyFilteredResults/4C_vs_0P_filtered.csv...\n",
      "Finished writing results to chea3Results/4C_vs_0P.tsv\n",
      "Generating TSV for pyFilteredResults/2P_vs_0P_filtered.csv...\n",
      "Finished writing results to chea3Results/2P_vs_0P.tsv\n",
      "Generating TSV for pyFilteredResults/1A_vs_0M_filtered.csv...\n",
      "Finished writing results to chea3Results/1A_vs_0M.tsv\n",
      "Generating TSV for pyFilteredResults/4C_vs_0M_filtered.csv...\n",
      "Finished writing results to chea3Results/4C_vs_0M.tsv\n",
      "Generating TSV for pyFilteredResults/3L_vs_0P_filtered.csv...\n",
      "Finished writing results to chea3Results/3L_vs_0P.tsv\n"
     ]
    }
   ],
   "source": [
    "# Run ChEA3 on each dataframe\n",
    "\n",
    "import csv\n",
    "import requests\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# The number of genes in the CSV required for ChEA3.\n",
    "# (We don't want to make a request to ChEA3 with an empty gene set,\n",
    "# for example.)\n",
    "thresh = 2\n",
    "\n",
    "src_dir = \"pyFilteredResults\"\n",
    "dst_dir = \"chea3Results\"\n",
    "\n",
    "# load the gene names from a CSV file.\n",
    "def get_gene_set(filename) -> list:\n",
    "    data = pd.read_csv(filename)\n",
    "\n",
    "    # return symbol column\n",
    "    return list(symbol for symbol in data[\"SYMBOL\"] if isinstance(symbol, str))\n",
    "\n",
    "# Creates a CSV file with ChEA3 using the given differentailly expressed\n",
    "# gene set and output filename.\n",
    "def make_chea3_tsv(gene_set, output_filename):\n",
    "    if len(gene_set) < thresh:\n",
    "        print(f\"Gene set only has {len(gene_set)} genes. Not producing CSV.\")\n",
    "        return\n",
    "\n",
    "    # The ChEA3 API URL.\n",
    "    url = \"https://maayanlab.cloud/chea3/api/enrich/\"\n",
    "\n",
    "    # Sends a JSON query to the ChEA3 API.\n",
    "    # https://maayanlab.cloud/chea3/\n",
    "    query_data = {\n",
    "        \"query_name\":\"myQuery\",\n",
    "        \"gene_set\":gene_set\n",
    "    }\n",
    "    result = requests.post(url, json=query_data)\n",
    "    assert result.ok, f\"Request failed with error code: {result.status_code}.\\\n",
    "    Check the ChEA3 website to make sure the request was correct.\"\n",
    "\n",
    "    # ChEA3 produces several tables (objects) in its JSON response.\n",
    "    # we get items from the table with this name:\n",
    "    entry = \"ENCODE--ChIP-seq\"\n",
    "    tfs = result.json()[entry]\n",
    "\n",
    "    with open(output_filename, \"w\") as tsvfile:\n",
    "        # Each item in the tfs is a dict. We make the table header using its keys.\n",
    "        fieldnames = tfs[0].keys()\n",
    "\n",
    "        writer = csv.DictWriter(tsvfile, fieldnames, delimiter=\"\\t\")\n",
    "        writer.writeheader()\n",
    "        for row in tfs:\n",
    "            writer.writerow(row)\n",
    "\n",
    "    print(f\"Finished writing results to {output_filename}\")\n",
    "\n",
    "!mkdir -p {dst_dir}\n",
    "\n",
    "for filename in os.listdir(src_dir):\n",
    "    print(f\"Generating TSV for {src_dir}/{filename}...\")\n",
    "    gene_set = get_gene_set(f\"{src_dir}/{filename}\")\n",
    "    new_filename = filename.rstrip(\"_filtered.csv\") + \".tsv\"\n",
    "    make_chea3_tsv(gene_set, f\"{dst_dir}/{new_filename}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Graphs\n",
    "Uses modified R code from a previous project to create graphs for each control-treatment pair.\n",
    "\n",
    "Input dir: chea3Results (`src_dir` variable)\n",
    "Output dir: graphs (`dst_dir` variable)\n",
    "\n",
    "Author: Randall J. Eck wrote transcriptionFactors.R. Slightly modified by Sky Duryee."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating graph for 4C_vs_0P.tsv...\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hchea3Results/4C_vs_0P.tsv graphs/4C_vs_0P.png\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"Reading from chea3Results/4C_vs_0P.tsv printing to graphs/4C_vs_0P.png\"\n",
      "\u001b[?25h\u001b[1mRows: \u001b[22m\u001b[34m118\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m12\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): Query Name, Set_name, TF, Library, Overlapping_Genes\n",
      "\u001b[32mdbl\u001b[39m (7): Rank, Scaled Rank, Intersect, Set length, FET p-value, FDR, Odds Ratio\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hGenerating graph for 3L_vs_0M.tsv...\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hchea3Results/3L_vs_0M.tsv graphs/3L_vs_0M.png\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"Reading from chea3Results/3L_vs_0M.tsv printing to graphs/3L_vs_0M.png\"\n",
      "\u001b[?25h\u001b[1mRows: \u001b[22m\u001b[34m118\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m12\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): Query Name, Set_name, TF, Library, Overlapping_Genes\n",
      "\u001b[32mdbl\u001b[39m (7): Rank, Scaled Rank, Intersect, Set length, FET p-value, FDR, Odds Ratio\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hGenerating graph for 3L_vs_0P.tsv...\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hchea3Results/3L_vs_0P.tsv graphs/3L_vs_0P.png\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"Reading from chea3Results/3L_vs_0P.tsv printing to graphs/3L_vs_0P.png\"\n",
      "\u001b[?25h\u001b[1mRows: \u001b[22m\u001b[34m118\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m12\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): Query Name, Set_name, TF, Library, Overlapping_Genes\n",
      "\u001b[32mdbl\u001b[39m (7): Rank, Scaled Rank, Intersect, Set length, FET p-value, FDR, Odds Ratio\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hGenerating graph for 4C_vs_0M.tsv...\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hchea3Results/4C_vs_0M.tsv graphs/4C_vs_0M.png\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"Reading from chea3Results/4C_vs_0M.tsv printing to graphs/4C_vs_0M.png\"\n",
      "\u001b[?25h\u001b[1mRows: \u001b[22m\u001b[34m118\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m12\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): Query Name, Set_name, TF, Library, Overlapping_Genes\n",
      "\u001b[32mdbl\u001b[39m (7): Rank, Scaled Rank, Intersect, Set length, FET p-value, FDR, Odds Ratio\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hGenerating graph for 1A_vs_0M.tsv...\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hchea3Results/1A_vs_0M.tsv graphs/1A_vs_0M.png\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"Reading from chea3Results/1A_vs_0M.tsv printing to graphs/1A_vs_0M.png\"\n",
      "\u001b[?25h\u001b[1mRows: \u001b[22m\u001b[34m118\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m12\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): Query Name, Set_name, TF, Library, Overlapping_Genes\n",
      "\u001b[32mdbl\u001b[39m (7): Rank, Scaled Rank, Intersect, Set length, FET p-value, FDR, Odds Ratio\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hGenerating graph for 1A_vs_0P.tsv...\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hchea3Results/1A_vs_0P.tsv graphs/1A_vs_0P.png\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"Reading from chea3Results/1A_vs_0P.tsv printing to graphs/1A_vs_0P.png\"\n",
      "\u001b[?25h\u001b[1mRows: \u001b[22m\u001b[34m118\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m12\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): Query Name, Set_name, TF, Library, Overlapping_Genes\n",
      "\u001b[32mdbl\u001b[39m (7): Rank, Scaled Rank, Intersect, Set length, FET p-value, FDR, Odds Ratio\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hGenerating graph for 2P_vs_0P.tsv...\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hchea3Results/2P_vs_0P.tsv graphs/2P_vs_0P.png\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"Reading from chea3Results/2P_vs_0P.tsv printing to graphs/2P_vs_0P.png\"\n",
      "\u001b[?25h\u001b[1mRows: \u001b[22m\u001b[34m118\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m12\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (5): Query Name, Set_name, TF, Library, Overlapping_Genes\n",
      "\u001b[32mdbl\u001b[39m (7): Rank, Scaled Rank, Intersect, Set length, FET p-value, FDR, Odds Ratio\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "src_dir = \"chea3Results\"\n",
    "dst_dir = \"graphs\"\n",
    "\n",
    "!mkdir -p graphs\n",
    "\n",
    "for file in os.listdir(src_dir):\n",
    "    print(f\"Generating graph for {file}...\")\n",
    "    assert file.endswith(\".tsv\"), f\"Only TSV files should be in the\\\n",
    "     {src_dir} folder\"\n",
    "\n",
    "    base = os.path.basename(file).rstrip(\".tsv\")\n",
    "\n",
    "    !Rscript transcriptionFactors.R {src_dir}/{file} {dst_dir}/{base}.png"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
